// @patternfly/react-core/dist/esm/components/EmptyState/EmptyStateHeader.js
import { __rest } from "tslib";
import * as React from 'react';
import { css } from '@patternfly/react-styles';
import styles from '@patternfly/react-styles/css/components/EmptyState/empty-state.mjs';
import { EmptyStateIcon } from './EmptyStateIcon';
export var EmptyStateHeadingLevel;
(function (EmptyStateHeadingLevel) {
    EmptyStateHeadingLevel["h1"] = "h1";
    EmptyStateHeadingLevel["h2"] = "h2";
    EmptyStateHeadingLevel["h3"] = "h3";
    EmptyStateHeadingLevel["h4"] = "h4";
    EmptyStateHeadingLevel["h5"] = "h5";
    EmptyStateHeadingLevel["h6"] = "h6";
})(EmptyStateHeadingLevel || (EmptyStateHeadingLevel = {}));
export const EmptyStateHeader = (_a) => {
    var { className, titleClassName, titleText, headingLevel: HeadingLevel = EmptyStateHeadingLevel.h1, icon: Icon } = _a, props = __rest(_a, ["className", "titleClassName", "titleText", "headingLevel", "icon"]);
    return (React.createElement("div", Object.assign({ className: css(`${styles.emptyState}__header`, className) }, props),
        Icon && React.createElement(EmptyStateIcon, { icon: Icon }),
        React.createElement("div", { className: css(`${styles.emptyState}__title`) },
            React.createElement(HeadingLevel, { className: css(styles.emptyStateTitleText, titleClassName) }, titleText))));
};
EmptyStateHeader.displayName = 'EmptyStateHeader';
