// @patternfly/react-core/dist/js/components/Content/Content.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Text = exports.TextVariants = void 0;
const tslib_1 = require("tslib");
const React = tslib_1.__importStar(require("react"));
const react_styles_1 = require("@patternfly/react-styles");
const content_1 = tslib_1.__importDefault(require("@patternfly/react-styles/css/components/Content/content"));
const helpers_1 = require("../../helpers");
var TextVariants;
(function (TextVariants) {
    TextVariants["h1"] = "h1";
    TextVariants["h2"] = "h2";
    TextVariants["h3"] = "h3";
    TextVariants["h4"] = "h4";
    TextVariants["h5"] = "h5";
    TextVariants["h6"] = "h6";
    TextVariants["p"] = "p";
    TextVariants["a"] = "a";
    TextVariants["small"] = "small";
    TextVariants["blockquote"] = "blockquote";
    TextVariants["pre"] = "pre";
    TextVariants["hr"] = "hr";
    TextVariants["ul"] = "ul";
    TextVariants["ol"] = "ol";
    TextVariants["dl"] = "dl";
    TextVariants["li"] = "li";
    TextVariants["dt"] = "dt";
    TextVariants["dd"] = "dd";
})(TextVariants || (exports.TextVariants = TextVariants = {}));
const componentStyles = {
    div: content_1.default.content,
    h1: content_1.default.contentH1,
    h2: content_1.default.contentH2,
    h3: content_1.default.contentH3,
    h4: content_1.default.contentH4,
    h5: content_1.default.contentH5,
    h6: content_1.default.contentH6,
    p: content_1.default.contentP,
    a: content_1.default.contentA,
    small: content_1.default.contentSmall,
    blockquote: content_1.default.contentBlockquote,
    pre: content_1.default.contentPre,
    hr: content_1.default.contentHr,
    ul: content_1.default.contentUl,
    ol: content_1.default.contentOl,
    dl: content_1.default.contentDl,
    li: content_1.default.contentLi,
    dt: content_1.default.contentDt,
    dd: content_1.default.contentDd
};
const Text = (_a) => {
    var { children, className = '', component, isPlainList = false, isVisitedLink = false, ouiaId, ouiaSafe = true } = _a, props = tslib_1.__rest(_a, ["children", "className", "component", "isPlainList", "isVisitedLink", "ouiaId", "ouiaSafe"]);
    const wrappingComponent = component !== null && component !== void 0 ? component : 'div';
    const Component = wrappingComponent;
    const ouiaProps = (0, helpers_1.useOUIAProps)(exports.Text.displayName, ouiaId, ouiaSafe);
    const isList = ['ul', 'ol', 'dl'].includes(wrappingComponent);
    return (React.createElement(Component, Object.assign({}, ouiaProps, props, { "data-pf-content": true, className: (0, react_styles_1.css)(componentStyles[wrappingComponent], isList && isPlainList && content_1.default.modifiers.plain, isVisitedLink && content_1.default.modifiers.visited, className) }), children));
};
exports.Text = Text;
exports.Text.displayName = 'Text';
