// @patternfly/react-topology/dist/esm/pipelines/components/anchors/TaskNodeSourceAnchor.js
import { Point } from '../../../geom';
import { AbstractAnchor } from '../../../anchors';
import { ScaleDetailsLevel } from '../../../types';
export default class TaskNodeSourceAnchor extends AbstractAnchor {
    constructor(owner, detailsLevel, lowDetailsStatusIconSize, vertical = false) {
        super(owner);
        this.lowDetailsStatusIconSize = 0;
        this.vertical = false;
        this.detailsLevel = detailsLevel;
        this.lowDetailsStatusIconSize = lowDetailsStatusIconSize;
        this.vertical = vertical;
    }
    getLocation() {
        return this.getReferencePoint();
    }
    getReferencePoint() {
        const bounds = this.owner.getBounds();
        if (this.detailsLevel !== ScaleDetailsLevel.high) {
            const nodeSize = this.lowDetailsStatusIconSize / this.owner.getGraph().getScale();
            if (this.vertical) {
                return new Point(bounds.x + bounds.width / 2, bounds.y + nodeSize);
            }
            return new Point(bounds.x + nodeSize, bounds.y + bounds.height / 2);
        }
        if (this.vertical) {
            return new Point(bounds.x + bounds.width / 2, bounds.bottom());
        }
        return new Point(bounds.right(), bounds.y + bounds.height / 2);
    }
}
