// @patternfly/react-topology/dist/esm/pipelines/components/anchors/TaskNodeTargetAnchor.js
import { Point } from '../../../geom';
import { AbstractAnchor } from '../../../anchors';
import { ScaleDetailsLevel } from '../../../types';
export default class TaskNodeTargetAnchor extends AbstractAnchor {
    constructor(owner, whenOffset, detailsLevel = ScaleDetailsLevel.high, lowDetailsStatusIconSize = 0, vertical = false) {
        super(owner);
        this.whenOffset = 0;
        this.lowDetailsStatusIconSize = 0;
        this.vertical = false;
        this.whenOffset = whenOffset;
        this.detailsLevel = detailsLevel;
        this.lowDetailsStatusIconSize = lowDetailsStatusIconSize;
        this.vertical = vertical;
    }
    getLocation() {
        return this.getReferencePoint();
    }
    getReferencePoint() {
        const bounds = this.owner.getBounds();
        if (this.vertical) {
            if (this.detailsLevel !== ScaleDetailsLevel.high) {
                return new Point(bounds.x + bounds.width / 2, bounds.y);
            }
            return new Point(bounds.x + bounds.width / 2, bounds.y - this.whenOffset);
        }
        return new Point(bounds.x, bounds.y + bounds.height / 2);
    }
}
