// @patternfly/react-topology/dist/js/pipelines/components/anchors/TaskNodeTargetAnchor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const geom_1 = require("../../../geom");
const anchors_1 = require("../../../anchors");
const types_1 = require("../../../types");
class TaskNodeTargetAnchor extends anchors_1.AbstractAnchor {
    constructor(owner, whenOffset, detailsLevel = types_1.ScaleDetailsLevel.high, lowDetailsStatusIconSize = 0, vertical = false) {
        super(owner);
        this.whenOffset = 0;
        this.lowDetailsStatusIconSize = 0;
        this.vertical = false;
        this.whenOffset = whenOffset;
        this.detailsLevel = detailsLevel;
        this.lowDetailsStatusIconSize = lowDetailsStatusIconSize;
        this.vertical = vertical;
    }
    getLocation() {
        return this.getReferencePoint();
    }
    getReferencePoint() {
        const bounds = this.owner.getBounds();
        if (this.vertical) {
            if (this.detailsLevel !== types_1.ScaleDetailsLevel.high) {
                return new geom_1.Point(bounds.x + bounds.width / 2, bounds.y);
            }
            return new geom_1.Point(bounds.x + bounds.width / 2, bounds.y - this.whenOffset);
        }
        return new geom_1.Point(bounds.x, bounds.y + bounds.height / 2);
    }
}
exports.default = TaskNodeTargetAnchor;
