// @patternfly/react-topology/dist/js/pipelines/components/anchors/TaskNodeSourceAnchor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const geom_1 = require("../../../geom");
const anchors_1 = require("../../../anchors");
const types_1 = require("../../../types");
class TaskNodeSourceAnchor extends anchors_1.AbstractAnchor {
    constructor(owner, detailsLevel, lowDetailsStatusIconSize, vertical = false) {
        super(owner);
        this.lowDetailsStatusIconSize = 0;
        this.vertical = false;
        this.detailsLevel = detailsLevel;
        this.lowDetailsStatusIconSize = lowDetailsStatusIconSize;
        this.vertical = vertical;
    }
    getLocation() {
        return this.getReferencePoint();
    }
    getReferencePoint() {
        const bounds = this.owner.getBounds();
        if (this.detailsLevel !== types_1.ScaleDetailsLevel.high) {
            const nodeSize = this.lowDetailsStatusIconSize / this.owner.getGraph().getScale();
            if (this.vertical) {
                return new geom_1.Point(bounds.x + bounds.width / 2, bounds.y + nodeSize);
            }
            return new geom_1.Point(bounds.x + nodeSize, bounds.y + bounds.height / 2);
        }
        if (this.vertical) {
            return new geom_1.Point(bounds.x + bounds.width / 2, bounds.bottom());
        }
        return new geom_1.Point(bounds.right(), bounds.y + bounds.height / 2);
    }
}
exports.default = TaskNodeSourceAnchor;
