// @storybook/core/dist/router/index.js
var ha = Object.create;
var Je = Object.defineProperty;
var da = Object.getOwnPropertyDescriptor;
var va = Object.getOwnPropertyNames;
var ma = Object.getPrototypeOf, ga = Object.prototype.hasOwnProperty;
var o = (t, e) => Je(t, "name", { value: e, configurable: !0 }), ce = /* @__PURE__ */ ((t) => typeof require < "u" ? require : typeof Proxy <
"u" ? new Proxy(t, {
  get: (e, r) => (typeof require < "u" ? require : e)[r]
}) : t)(function(t) {
  if (typeof require < "u") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + t + '" is not supported');
});
var b = (t, e) => () => (e || t((e = { exports: {} }).exports, e), e.exports);
var ba = (t, e, r, n) => {
  if (e && typeof e == "object" || typeof e == "function")
    for (let a of va(e))
      !ga.call(t, a) && a !== r && Je(t, a, { get: () => e[a], enumerable: !(n = da(e, a)) || n.enumerable });
  return t;
};
var Ke = (t, e, r) => (r = t != null ? ha(ma(t)) : {}, ba(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  e || !t || !t.__esModule ? Je(r, "default", { value: t, enumerable: !0 }) : r,
  t
));

// ../node_modules/lodash/_freeGlobal.js
var $t = b((Nl, Ft) => {
  var Sa = typeof global == "object" && global && global.Object === Object && global;
  Ft.exports = Sa;
});

// ../node_modules/lodash/_root.js
var kt = b((Il, Ct) => {
  var wa = $t(), Pa = typeof self == "object" && self && self.Object === Object && self, Ea = wa || Pa || Function("return this")();
  Ct.exports = Ea;
});

// ../node_modules/lodash/_Symbol.js
var Ye = b((_l, qt) => {
  var Oa = kt(), xa = Oa.Symbol;
  qt.exports = xa;
});

// ../node_modules/lodash/_getRawTag.js
var Vt = b((Dl, Ut) => {
  var Wt = Ye(), Bt = Object.prototype, Ra = Bt.hasOwnProperty, Aa = Bt.toString, fe = Wt ? Wt.toStringTag : void 0;
  function Na(t) {
    var e = Ra.call(t, fe), r = t[fe];
    try {
      t[fe] = void 0;
      var n = !0;
    } catch {
    }
    var a = Aa.call(t);
    return n && (e ? t[fe] = r : delete t[fe]), a;
  }
  o(Na, "getRawTag");
  Ut.exports = Na;
});

// ../node_modules/lodash/_objectToString.js
var Ht = b((Ml, jt) => {
  var Ia = Object.prototype, _a = Ia.toString;
  function Da(t) {
    return _a.call(t);
  }
  o(Da, "objectToString");
  jt.exports = Da;
});

// ../node_modules/lodash/_baseGetTag.js
var Jt = b((Fl, Qt) => {
  var zt = Ye(), La = Vt(), Ma = Ht(), Ta = "[object Null]", Fa = "[object Undefined]", Gt = zt ? zt.toStringTag : void 0;
  function $a(t) {
    return t == null ? t === void 0 ? Fa : Ta : Gt && Gt in Object(t) ? La(t) : Ma(t);
  }
  o($a, "baseGetTag");
  Qt.exports = $a;
});

// ../node_modules/lodash/_overArg.js
var Yt = b((Cl, Kt) => {
  function Ca(t, e) {
    return function(r) {
      return t(e(r));
    };
  }
  o(Ca, "overArg");
  Kt.exports = Ca;
});

// ../node_modules/lodash/_getPrototype.js
var Zt = b((ql, Xt) => {
  var ka = Yt(), qa = ka(Object.getPrototypeOf, Object);
  Xt.exports = qa;
});

// ../node_modules/lodash/isObjectLike.js
var tr = b((Wl, er) => {
  function Wa(t) {
    return t != null && typeof t == "object";
  }
  o(Wa, "isObjectLike");
  er.exports = Wa;
});

// ../node_modules/lodash/isPlainObject.js
var ar = b((Ul, nr) => {
  var Ba = Jt(), Ua = Zt(), Va = tr(), ja = "[object Object]", Ha = Function.prototype, za = Object.prototype, rr = Ha.toString, Ga = za.hasOwnProperty,
  Qa = rr.call(Object);
  function Ja(t) {
    if (!Va(t) || Ba(t) != ja)
      return !1;
    var e = Ua(t);
    if (e === null)
      return !0;
    var r = Ga.call(e, "constructor") && e.constructor;
    return typeof r == "function" && r instanceof r && rr.call(r) == Qa;
  }
  o(Ja, "isPlainObject");
  nr.exports = Ja;
});

// ../node_modules/memoizerific/memoizerific.js
var ir = b((or, Xe) => {
  (function(t) {
    if (typeof or == "object" && typeof Xe < "u")
      Xe.exports = t();
    else if (typeof define == "function" && define.amd)
      define([], t);
    else {
      var e;
      typeof window < "u" ? e = window : typeof global < "u" ? e = global : typeof self < "u" ? e = self : e = this, e.memoizerific = t();
    }
  })(function() {
    var t, e, r;
    return (/* @__PURE__ */ o(function n(a, i, s) {
      function l(f, h) {
        if (!i[f]) {
          if (!a[f]) {
            var p = typeof ce == "function" && ce;
            if (!h && p) return p(f, !0);
            if (u) return u(f, !0);
            var d = new Error("Cannot find module '" + f + "'");
            throw d.code = "MODULE_NOT_FOUND", d;
          }
          var y = i[f] = { exports: {} };
          a[f][0].call(y.exports, function(v) {
            var P = a[f][1][v];
            return l(P || v);
          }, y, y.exports, n, a, i, s);
        }
        return i[f].exports;
      }
      o(l, "s");
      for (var u = typeof ce == "function" && ce, c = 0; c < s.length; c++) l(s[c]);
      return l;
    }, "e"))({ 1: [function(n, a, i) {
      a.exports = function(s) {
        if (typeof Map != "function" || s) {
          var l = n("./similar");
          return new l();
        } else
          return /* @__PURE__ */ new Map();
      };
    }, { "./similar": 2 }], 2: [function(n, a, i) {
      function s() {
        return this.list = [], this.lastItem = void 0, this.size = 0, this;
      }
      o(s, "Similar"), s.prototype.get = function(l) {
        var u;
        if (this.lastItem && this.isEqual(this.lastItem.key, l))
          return this.lastItem.val;
        if (u = this.indexOf(l), u >= 0)
          return this.lastItem = this.list[u], this.list[u].val;
      }, s.prototype.set = function(l, u) {
        var c;
        return this.lastItem && this.isEqual(this.lastItem.key, l) ? (this.lastItem.val = u, this) : (c = this.indexOf(l), c >= 0 ? (this.lastItem =
        this.list[c], this.list[c].val = u, this) : (this.lastItem = { key: l, val: u }, this.list.push(this.lastItem), this.size++, this));
      }, s.prototype.delete = function(l) {
        var u;
        if (this.lastItem && this.isEqual(this.lastItem.key, l) && (this.lastItem = void 0), u = this.indexOf(l), u >= 0)
          return this.size--, this.list.splice(u, 1)[0];
      }, s.prototype.has = function(l) {
        var u;
        return this.lastItem && this.isEqual(this.lastItem.key, l) ? !0 : (u = this.indexOf(l), u >= 0 ? (this.lastItem = this.list[u], !0) :
        !1);
      }, s.prototype.forEach = function(l, u) {
        var c;
        for (c = 0; c < this.size; c++)
          l.call(u || this, this.list[c].val, this.list[c].key, this);
      }, s.prototype.indexOf = function(l) {
        var u;
        for (u = 0; u < this.size; u++)
          if (this.isEqual(this.list[u].key, l))
            return u;
        return -1;
      }, s.prototype.isEqual = function(l, u) {
        return l === u || l !== l && u !== u;
      }, a.exports = s;
    }, {}], 3: [function(n, a, i) {
      var s = n("map-or-similar");
      a.exports = function(f) {
        var h = new s(!1), p = [];
        return function(d) {
          var y = /* @__PURE__ */ o(function() {
            var v = h, P, g, E = arguments.length - 1, _ = Array(E + 1), D = !0, x;
            if ((y.numArgs || y.numArgs === 0) && y.numArgs !== E + 1)
              throw new Error("Memoizerific functions should always be called with the same number of arguments");
            for (x = 0; x < E; x++) {
              if (_[x] = {
                cacheItem: v,
                arg: arguments[x]
              }, v.has(arguments[x])) {
                v = v.get(arguments[x]);
                continue;
              }
              D = !1, P = new s(!1), v.set(arguments[x], P), v = P;
            }
            return D && (v.has(arguments[E]) ? g = v.get(arguments[E]) : D = !1), D || (g = d.apply(null, arguments), v.set(arguments[E], g)),
            f > 0 && (_[E] = {
              cacheItem: v,
              arg: arguments[E]
            }, D ? l(p, _) : p.push(_), p.length > f && u(p.shift())), y.wasMemoized = D, y.numArgs = E + 1, g;
          }, "memoizerific");
          return y.limit = f, y.wasMemoized = !1, y.cache = h, y.lru = p, y;
        };
      };
      function l(f, h) {
        var p = f.length, d = h.length, y, v, P;
        for (v = 0; v < p; v++) {
          for (y = !0, P = 0; P < d; P++)
            if (!c(f[v][P].arg, h[P].arg)) {
              y = !1;
              break;
            }
          if (y)
            break;
        }
        f.push(f.splice(v, 1)[0]);
      }
      o(l, "moveToMostRecentLru");
      function u(f) {
        var h = f.length, p = f[h - 1], d, y;
        for (p.cacheItem.delete(p.arg), y = h - 2; y >= 0 && (p = f[y], d = p.cacheItem.get(p.arg), !d || !d.size); y--)
          p.cacheItem.delete(p.arg);
      }
      o(u, "removeCachedResult");
      function c(f, h) {
        return f === h || f !== f && h !== h;
      }
      o(c, "isEqual");
    }, { "map-or-similar": 1 }] }, {}, [3])(3);
  });
});

// ../node_modules/es-errors/index.js
var sr = b((Hl, lr) => {
  "use strict";
  lr.exports = Error;
});

// ../node_modules/es-errors/eval.js
var cr = b((zl, ur) => {
  "use strict";
  ur.exports = EvalError;
});

// ../node_modules/es-errors/range.js
var pr = b((Gl, fr) => {
  "use strict";
  fr.exports = RangeError;
});

// ../node_modules/es-errors/ref.js
var hr = b((Ql, yr) => {
  "use strict";
  yr.exports = ReferenceError;
});

// ../node_modules/es-errors/syntax.js
var Ze = b((Jl, dr) => {
  "use strict";
  dr.exports = SyntaxError;
});

// ../node_modules/es-errors/type.js
var pe = b((Kl, vr) => {
  "use strict";
  vr.exports = TypeError;
});

// ../node_modules/es-errors/uri.js
var gr = b((Yl, mr) => {
  "use strict";
  mr.exports = URIError;
});

// ../node_modules/has-symbols/shams.js
var Sr = b((Xl, br) => {
  "use strict";
  br.exports = /* @__PURE__ */ o(function() {
    if (typeof Symbol != "function" || typeof Object.getOwnPropertySymbols != "function")
      return !1;
    if (typeof Symbol.iterator == "symbol")
      return !0;
    var e = {}, r = Symbol("test"), n = Object(r);
    if (typeof r == "string" || Object.prototype.toString.call(r) !== "[object Symbol]" || Object.prototype.toString.call(n) !== "[object Sy\
mbol]")
      return !1;
    var a = 42;
    e[r] = a;
    for (r in e)
      return !1;
    if (typeof Object.keys == "function" && Object.keys(e).length !== 0 || typeof Object.getOwnPropertyNames == "function" && Object.getOwnPropertyNames(
    e).length !== 0)
      return !1;
    var i = Object.getOwnPropertySymbols(e);
    if (i.length !== 1 || i[0] !== r || !Object.prototype.propertyIsEnumerable.call(e, r))
      return !1;
    if (typeof Object.getOwnPropertyDescriptor == "function") {
      var s = Object.getOwnPropertyDescriptor(e, r);
      if (s.value !== a || s.enumerable !== !0)
        return !1;
    }
    return !0;
  }, "hasSymbols");
});

// ../node_modules/has-symbols/index.js
var Er = b((es, Pr) => {
  "use strict";
  var wr = typeof Symbol < "u" && Symbol, Ka = Sr();
  Pr.exports = /* @__PURE__ */ o(function() {
    return typeof wr != "function" || typeof Symbol != "function" || typeof wr("foo") != "symbol" || typeof Symbol("bar") != "symbol" ? !1 :
    Ka();
  }, "hasNativeSymbols");
});

// ../node_modules/has-proto/index.js
var xr = b((rs, Or) => {
  "use strict";
  var et = {
    __proto__: null,
    foo: {}
  }, Ya = Object;
  Or.exports = /* @__PURE__ */ o(function() {
    return { __proto__: et }.foo === et.foo && !(et instanceof Ya);
  }, "hasProto");
});

// ../node_modules/function-bind/implementation.js
var Nr = b((as, Ar) => {
  "use strict";
  var Xa = "Function.prototype.bind called on incompatible ", Za = Object.prototype.toString, eo = Math.max, to = "[object Function]", Rr = /* @__PURE__ */ o(
  function(e, r) {
    for (var n = [], a = 0; a < e.length; a += 1)
      n[a] = e[a];
    for (var i = 0; i < r.length; i += 1)
      n[i + e.length] = r[i];
    return n;
  }, "concatty"), ro = /* @__PURE__ */ o(function(e, r) {
    for (var n = [], a = r || 0, i = 0; a < e.length; a += 1, i += 1)
      n[i] = e[a];
    return n;
  }, "slicy"), no = /* @__PURE__ */ o(function(t, e) {
    for (var r = "", n = 0; n < t.length; n += 1)
      r += t[n], n + 1 < t.length && (r += e);
    return r;
  }, "joiny");
  Ar.exports = /* @__PURE__ */ o(function(e) {
    var r = this;
    if (typeof r != "function" || Za.apply(r) !== to)
      throw new TypeError(Xa + r);
    for (var n = ro(arguments, 1), a, i = /* @__PURE__ */ o(function() {
      if (this instanceof a) {
        var f = r.apply(
          this,
          Rr(n, arguments)
        );
        return Object(f) === f ? f : this;
      }
      return r.apply(
        e,
        Rr(n, arguments)
      );
    }, "binder"), s = eo(0, r.length - n.length), l = [], u = 0; u < s; u++)
      l[u] = "$" + u;
    if (a = Function("binder", "return function (" + no(l, ",") + "){ return binder.apply(this,arguments); }")(i), r.prototype) {
      var c = /* @__PURE__ */ o(function() {
      }, "Empty");
      c.prototype = r.prototype, a.prototype = new c(), c.prototype = null;
    }
    return a;
  }, "bind");
});

// ../node_modules/function-bind/index.js
var Re = b((is, Ir) => {
  "use strict";
  var ao = Nr();
  Ir.exports = Function.prototype.bind || ao;
});

// ../node_modules/hasown/index.js
var Dr = b((ls, _r) => {
  "use strict";
  var oo = Function.prototype.call, io = Object.prototype.hasOwnProperty, lo = Re();
  _r.exports = lo.call(oo, io);
});

// ../node_modules/get-intrinsic/index.js
var G = b((ss, $r) => {
  "use strict";
  var S, so = sr(), uo = cr(), co = pr(), fo = hr(), te = Ze(), ee = pe(), po = gr(), Fr = Function, tt = /* @__PURE__ */ o(function(t) {
    try {
      return Fr('"use strict"; return (' + t + ").constructor;")();
    } catch {
    }
  }, "getEvalledConstructor"), H = Object.getOwnPropertyDescriptor;
  if (H)
    try {
      H({}, "");
    } catch {
      H = null;
    }
  var rt = /* @__PURE__ */ o(function() {
    throw new ee();
  }, "throwTypeError"), yo = H ? function() {
    try {
      return arguments.callee, rt;
    } catch {
      try {
        return H(arguments, "callee").get;
      } catch {
        return rt;
      }
    }
  }() : rt, X = Er()(), ho = xr()(), A = Object.getPrototypeOf || (ho ? function(t) {
    return t.__proto__;
  } : null), Z = {}, vo = typeof Uint8Array > "u" || !A ? S : A(Uint8Array), z = {
    __proto__: null,
    "%AggregateError%": typeof AggregateError > "u" ? S : AggregateError,
    "%Array%": Array,
    "%ArrayBuffer%": typeof ArrayBuffer > "u" ? S : ArrayBuffer,
    "%ArrayIteratorPrototype%": X && A ? A([][Symbol.iterator]()) : S,
    "%AsyncFromSyncIteratorPrototype%": S,
    "%AsyncFunction%": Z,
    "%AsyncGenerator%": Z,
    "%AsyncGeneratorFunction%": Z,
    "%AsyncIteratorPrototype%": Z,
    "%Atomics%": typeof Atomics > "u" ? S : Atomics,
    "%BigInt%": typeof BigInt > "u" ? S : BigInt,
    "%BigInt64Array%": typeof BigInt64Array > "u" ? S : BigInt64Array,
    "%BigUint64Array%": typeof BigUint64Array > "u" ? S : BigUint64Array,
    "%Boolean%": Boolean,
    "%DataView%": typeof DataView > "u" ? S : DataView,
    "%Date%": Date,
    "%decodeURI%": decodeURI,
    "%decodeURIComponent%": decodeURIComponent,
    "%encodeURI%": encodeURI,
    "%encodeURIComponent%": encodeURIComponent,
    "%Error%": so,
    "%eval%": eval,
    // eslint-disable-line no-eval
    "%EvalError%": uo,
    "%Float32Array%": typeof Float32Array > "u" ? S : Float32Array,
    "%Float64Array%": typeof Float64Array > "u" ? S : Float64Array,
    "%FinalizationRegistry%": typeof FinalizationRegistry > "u" ? S : FinalizationRegistry,
    "%Function%": Fr,
    "%GeneratorFunction%": Z,
    "%Int8Array%": typeof Int8Array > "u" ? S : Int8Array,
    "%Int16Array%": typeof Int16Array > "u" ? S : Int16Array,
    "%Int32Array%": typeof Int32Array > "u" ? S : Int32Array,
    "%isFinite%": isFinite,
    "%isNaN%": isNaN,
    "%IteratorPrototype%": X && A ? A(A([][Symbol.iterator]())) : S,
    "%JSON%": typeof JSON == "object" ? JSON : S,
    "%Map%": typeof Map > "u" ? S : Map,
    "%MapIteratorPrototype%": typeof Map > "u" || !X || !A ? S : A((/* @__PURE__ */ new Map())[Symbol.iterator]()),
    "%Math%": Math,
    "%Number%": Number,
    "%Object%": Object,
    "%parseFloat%": parseFloat,
    "%parseInt%": parseInt,
    "%Promise%": typeof Promise > "u" ? S : Promise,
    "%Proxy%": typeof Proxy > "u" ? S : Proxy,
    "%RangeError%": co,
    "%ReferenceError%": fo,
    "%Reflect%": typeof Reflect > "u" ? S : Reflect,
    "%RegExp%": RegExp,
    "%Set%": typeof Set > "u" ? S : Set,
    "%SetIteratorPrototype%": typeof Set > "u" || !X || !A ? S : A((/* @__PURE__ */ new Set())[Symbol.iterator]()),
    "%SharedArrayBuffer%": typeof SharedArrayBuffer > "u" ? S : SharedArrayBuffer,
    "%String%": String,
    "%StringIteratorPrototype%": X && A ? A(""[Symbol.iterator]()) : S,
    "%Symbol%": X ? Symbol : S,
    "%SyntaxError%": te,
    "%ThrowTypeError%": yo,
    "%TypedArray%": vo,
    "%TypeError%": ee,
    "%Uint8Array%": typeof Uint8Array > "u" ? S : Uint8Array,
    "%Uint8ClampedArray%": typeof Uint8ClampedArray > "u" ? S : Uint8ClampedArray,
    "%Uint16Array%": typeof Uint16Array > "u" ? S : Uint16Array,
    "%Uint32Array%": typeof Uint32Array > "u" ? S : Uint32Array,
    "%URIError%": po,
    "%WeakMap%": typeof WeakMap > "u" ? S : WeakMap,
    "%WeakRef%": typeof WeakRef > "u" ? S : WeakRef,
    "%WeakSet%": typeof WeakSet > "u" ? S : WeakSet
  };
  if (A)
    try {
      null.error;
    } catch (t) {
      Lr = A(A(t)), z["%Error.prototype%"] = Lr;
    }
  var Lr, mo = /* @__PURE__ */ o(function t(e) {
    var r;
    if (e === "%AsyncFunction%")
      r = tt("async function () {}");
    else if (e === "%GeneratorFunction%")
      r = tt("function* () {}");
    else if (e === "%AsyncGeneratorFunction%")
      r = tt("async function* () {}");
    else if (e === "%AsyncGenerator%") {
      var n = t("%AsyncGeneratorFunction%");
      n && (r = n.prototype);
    } else if (e === "%AsyncIteratorPrototype%") {
      var a = t("%AsyncGenerator%");
      a && A && (r = A(a.prototype));
    }
    return z[e] = r, r;
  }, "doEval"), Mr = {
    __proto__: null,
    "%ArrayBufferPrototype%": ["ArrayBuffer", "prototype"],
    "%ArrayPrototype%": ["Array", "prototype"],
    "%ArrayProto_entries%": ["Array", "prototype", "entries"],
    "%ArrayProto_forEach%": ["Array", "prototype", "forEach"],
    "%ArrayProto_keys%": ["Array", "prototype", "keys"],
    "%ArrayProto_values%": ["Array", "prototype", "values"],
    "%AsyncFunctionPrototype%": ["AsyncFunction", "prototype"],
    "%AsyncGenerator%": ["AsyncGeneratorFunction", "prototype"],
    "%AsyncGeneratorPrototype%": ["AsyncGeneratorFunction", "prototype", "prototype"],
    "%BooleanPrototype%": ["Boolean", "prototype"],
    "%DataViewPrototype%": ["DataView", "prototype"],
    "%DatePrototype%": ["Date", "prototype"],
    "%ErrorPrototype%": ["Error", "prototype"],
    "%EvalErrorPrototype%": ["EvalError", "prototype"],
    "%Float32ArrayPrototype%": ["Float32Array", "prototype"],
    "%Float64ArrayPrototype%": ["Float64Array", "prototype"],
    "%FunctionPrototype%": ["Function", "prototype"],
    "%Generator%": ["GeneratorFunction", "prototype"],
    "%GeneratorPrototype%": ["GeneratorFunction", "prototype", "prototype"],
    "%Int8ArrayPrototype%": ["Int8Array", "prototype"],
    "%Int16ArrayPrototype%": ["Int16Array", "prototype"],
    "%Int32ArrayPrototype%": ["Int32Array", "prototype"],
    "%JSONParse%": ["JSON", "parse"],
    "%JSONStringify%": ["JSON", "stringify"],
    "%MapPrototype%": ["Map", "prototype"],
    "%NumberPrototype%": ["Number", "prototype"],
    "%ObjectPrototype%": ["Object", "prototype"],
    "%ObjProto_toString%": ["Object", "prototype", "toString"],
    "%ObjProto_valueOf%": ["Object", "prototype", "valueOf"],
    "%PromisePrototype%": ["Promise", "prototype"],
    "%PromiseProto_then%": ["Promise", "prototype", "then"],
    "%Promise_all%": ["Promise", "all"],
    "%Promise_reject%": ["Promise", "reject"],
    "%Promise_resolve%": ["Promise", "resolve"],
    "%RangeErrorPrototype%": ["RangeError", "prototype"],
    "%ReferenceErrorPrototype%": ["ReferenceError", "prototype"],
    "%RegExpPrototype%": ["RegExp", "prototype"],
    "%SetPrototype%": ["Set", "prototype"],
    "%SharedArrayBufferPrototype%": ["SharedArrayBuffer", "prototype"],
    "%StringPrototype%": ["String", "prototype"],
    "%SymbolPrototype%": ["Symbol", "prototype"],
    "%SyntaxErrorPrototype%": ["SyntaxError", "prototype"],
    "%TypedArrayPrototype%": ["TypedArray", "prototype"],
    "%TypeErrorPrototype%": ["TypeError", "prototype"],
    "%Uint8ArrayPrototype%": ["Uint8Array", "prototype"],
    "%Uint8ClampedArrayPrototype%": ["Uint8ClampedArray", "prototype"],
    "%Uint16ArrayPrototype%": ["Uint16Array", "prototype"],
    "%Uint32ArrayPrototype%": ["Uint32Array", "prototype"],
    "%URIErrorPrototype%": ["URIError", "prototype"],
    "%WeakMapPrototype%": ["WeakMap", "prototype"],
    "%WeakSetPrototype%": ["WeakSet", "prototype"]
  }, ye = Re(), Ae = Dr(), go = ye.call(Function.call, Array.prototype.concat), bo = ye.call(Function.apply, Array.prototype.splice), Tr = ye.
  call(Function.call, String.prototype.replace), Ne = ye.call(Function.call, String.prototype.slice), So = ye.call(Function.call, RegExp.prototype.
  exec), wo = /[^%.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|%$))/g, Po = /\\(\\)?/g, Eo = /* @__PURE__ */ o(
  function(e) {
    var r = Ne(e, 0, 1), n = Ne(e, -1);
    if (r === "%" && n !== "%")
      throw new te("invalid intrinsic syntax, expected closing `%`");
    if (n === "%" && r !== "%")
      throw new te("invalid intrinsic syntax, expected opening `%`");
    var a = [];
    return Tr(e, wo, function(i, s, l, u) {
      a[a.length] = l ? Tr(u, Po, "$1") : s || i;
    }), a;
  }, "stringToPath"), Oo = /* @__PURE__ */ o(function(e, r) {
    var n = e, a;
    if (Ae(Mr, n) && (a = Mr[n], n = "%" + a[0] + "%"), Ae(z, n)) {
      var i = z[n];
      if (i === Z && (i = mo(n)), typeof i > "u" && !r)
        throw new ee("intrinsic " + e + " exists, but is not available. Please file an issue!");
      return {
        alias: a,
        name: n,
        value: i
      };
    }
    throw new te("intrinsic " + e + " does not exist!");
  }, "getBaseIntrinsic");
  $r.exports = /* @__PURE__ */ o(function(e, r) {
    if (typeof e != "string" || e.length === 0)
      throw new ee("intrinsic name must be a non-empty string");
    if (arguments.length > 1 && typeof r != "boolean")
      throw new ee('"allowMissing" argument must be a boolean');
    if (So(/^%?[^%]*%?$/, e) === null)
      throw new te("`%` may not be present anywhere but at the beginning and end of the intrinsic name");
    var n = Eo(e), a = n.length > 0 ? n[0] : "", i = Oo("%" + a + "%", r), s = i.name, l = i.value, u = !1, c = i.alias;
    c && (a = c[0], bo(n, go([0, 1], c)));
    for (var f = 1, h = !0; f < n.length; f += 1) {
      var p = n[f], d = Ne(p, 0, 1), y = Ne(p, -1);
      if ((d === '"' || d === "'" || d === "`" || y === '"' || y === "'" || y === "`") && d !== y)
        throw new te("property names with quotes must have matching quotes");
      if ((p === "constructor" || !h) && (u = !0), a += "." + p, s = "%" + a + "%", Ae(z, s))
        l = z[s];
      else if (l != null) {
        if (!(p in l)) {
          if (!r)
            throw new ee("base intrinsic for " + e + " exists, but the property is not available.");
          return;
        }
        if (H && f + 1 >= n.length) {
          var v = H(l, p);
          h = !!v, h && "get" in v && !("originalValue" in v.get) ? l = v.get : l = l[p];
        } else
          h = Ae(l, p), l = l[p];
        h && !u && (z[s] = l);
      }
    }
    return l;
  }, "GetIntrinsic");
});

// ../node_modules/es-define-property/index.js
var _e = b((cs, Cr) => {
  "use strict";
  var xo = G(), Ie = xo("%Object.defineProperty%", !0) || !1;
  if (Ie)
    try {
      Ie({}, "a", { value: 1 });
    } catch {
      Ie = !1;
    }
  Cr.exports = Ie;
});

// ../node_modules/gopd/index.js
var nt = b((fs, kr) => {
  "use strict";
  var Ro = G(), De = Ro("%Object.getOwnPropertyDescriptor%", !0);
  if (De)
    try {
      De([], "length");
    } catch {
      De = null;
    }
  kr.exports = De;
});

// ../node_modules/define-data-property/index.js
var Ur = b((ps, Br) => {
  "use strict";
  var qr = _e(), Ao = Ze(), re = pe(), Wr = nt();
  Br.exports = /* @__PURE__ */ o(function(e, r, n) {
    if (!e || typeof e != "object" && typeof e != "function")
      throw new re("`obj` must be an object or a function`");
    if (typeof r != "string" && typeof r != "symbol")
      throw new re("`property` must be a string or a symbol`");
    if (arguments.length > 3 && typeof arguments[3] != "boolean" && arguments[3] !== null)
      throw new re("`nonEnumerable`, if provided, must be a boolean or null");
    if (arguments.length > 4 && typeof arguments[4] != "boolean" && arguments[4] !== null)
      throw new re("`nonWritable`, if provided, must be a boolean or null");
    if (arguments.length > 5 && typeof arguments[5] != "boolean" && arguments[5] !== null)
      throw new re("`nonConfigurable`, if provided, must be a boolean or null");
    if (arguments.length > 6 && typeof arguments[6] != "boolean")
      throw new re("`loose`, if provided, must be a boolean");
    var a = arguments.length > 3 ? arguments[3] : null, i = arguments.length > 4 ? arguments[4] : null, s = arguments.length > 5 ? arguments[5] :
    null, l = arguments.length > 6 ? arguments[6] : !1, u = !!Wr && Wr(e, r);
    if (qr)
      qr(e, r, {
        configurable: s === null && u ? u.configurable : !s,
        enumerable: a === null && u ? u.enumerable : !a,
        value: n,
        writable: i === null && u ? u.writable : !i
      });
    else if (l || !a && !i && !s)
      e[r] = n;
    else
      throw new Ao("This environment does not support defining a property as non-configurable, non-writable, or non-enumerable.");
  }, "defineDataProperty");
});

// ../node_modules/has-property-descriptors/index.js
var Hr = b((hs, jr) => {
  "use strict";
  var at = _e(), Vr = /* @__PURE__ */ o(function() {
    return !!at;
  }, "hasPropertyDescriptors");
  Vr.hasArrayLengthDefineBug = /* @__PURE__ */ o(function() {
    if (!at)
      return null;
    try {
      return at([], "length", { value: 1 }).length !== 1;
    } catch {
      return !0;
    }
  }, "hasArrayLengthDefineBug");
  jr.exports = Vr;
});

// ../node_modules/set-function-length/index.js
var Kr = b((vs, Jr) => {
  "use strict";
  var No = G(), zr = Ur(), Io = Hr()(), Gr = nt(), Qr = pe(), _o = No("%Math.floor%");
  Jr.exports = /* @__PURE__ */ o(function(e, r) {
    if (typeof e != "function")
      throw new Qr("`fn` is not a function");
    if (typeof r != "number" || r < 0 || r > 4294967295 || _o(r) !== r)
      throw new Qr("`length` must be a positive 32-bit integer");
    var n = arguments.length > 2 && !!arguments[2], a = !0, i = !0;
    if ("length" in e && Gr) {
      var s = Gr(e, "length");
      s && !s.configurable && (a = !1), s && !s.writable && (i = !1);
    }
    return (a || i || !n) && (Io ? zr(
      /** @type {Parameters<define>[0]} */
      e,
      "length",
      r,
      !0,
      !0
    ) : zr(
      /** @type {Parameters<define>[0]} */
      e,
      "length",
      r
    )), e;
  }, "setFunctionLength");
});

// ../node_modules/call-bind/index.js
var rn = b((gs, Le) => {
  "use strict";
  var ot = Re(), Me = G(), Do = Kr(), Lo = pe(), Zr = Me("%Function.prototype.apply%"), en = Me("%Function.prototype.call%"), tn = Me("%Refl\
ect.apply%", !0) || ot.call(en, Zr), Yr = _e(), Mo = Me("%Math.max%");
  Le.exports = /* @__PURE__ */ o(function(e) {
    if (typeof e != "function")
      throw new Lo("a function is required");
    var r = tn(ot, en, arguments);
    return Do(
      r,
      1 + Mo(0, e.length - (arguments.length - 1)),
      !0
    );
  }, "callBind");
  var Xr = /* @__PURE__ */ o(function() {
    return tn(ot, Zr, arguments);
  }, "applyBind");
  Yr ? Yr(Le.exports, "apply", { value: Xr }) : Le.exports.apply = Xr;
});

// ../node_modules/call-bind/callBound.js
var ln = b((Ss, on) => {
  "use strict";
  var nn = G(), an = rn(), To = an(nn("String.prototype.indexOf"));
  on.exports = /* @__PURE__ */ o(function(e, r) {
    var n = nn(e, !!r);
    return typeof n == "function" && To(e, ".prototype.") > -1 ? an(n) : n;
  }, "callBoundIntrinsic");
});

// (disabled):../node_modules/object-inspect/util.inspect
var sn = b(() => {
});

// ../node_modules/object-inspect/index.js
var Nn = b((Os, An) => {
  var dt = typeof Map == "function" && Map.prototype, it = Object.getOwnPropertyDescriptor && dt ? Object.getOwnPropertyDescriptor(Map.prototype,
  "size") : null, Fe = dt && it && typeof it.get == "function" ? it.get : null, un = dt && Map.prototype.forEach, vt = typeof Set == "functi\
on" && Set.prototype, lt = Object.getOwnPropertyDescriptor && vt ? Object.getOwnPropertyDescriptor(Set.prototype, "size") : null, $e = vt &&
  lt && typeof lt.get == "function" ? lt.get : null, cn = vt && Set.prototype.forEach, Fo = typeof WeakMap == "function" && WeakMap.prototype,
  de = Fo ? WeakMap.prototype.has : null, $o = typeof WeakSet == "function" && WeakSet.prototype, ve = $o ? WeakSet.prototype.has : null, Co = typeof WeakRef ==
  "function" && WeakRef.prototype, fn = Co ? WeakRef.prototype.deref : null, ko = Boolean.prototype.valueOf, qo = Object.prototype.toString,
  Wo = Function.prototype.toString, Bo = String.prototype.match, mt = String.prototype.slice, B = String.prototype.replace, Uo = String.prototype.
  toUpperCase, pn = String.prototype.toLowerCase, wn = RegExp.prototype.test, yn = Array.prototype.concat, C = Array.prototype.join, Vo = Array.
  prototype.slice, hn = Math.floor, ct = typeof BigInt == "function" ? BigInt.prototype.valueOf : null, st = Object.getOwnPropertySymbols, ft = typeof Symbol ==
  "function" && typeof Symbol.iterator == "symbol" ? Symbol.prototype.toString : null, ne = typeof Symbol == "function" && typeof Symbol.iterator ==
  "object", M = typeof Symbol == "function" && Symbol.toStringTag && (typeof Symbol.toStringTag === ne || !0) ? Symbol.toStringTag : null, Pn = Object.
  prototype.propertyIsEnumerable, dn = (typeof Reflect == "function" ? Reflect.getPrototypeOf : Object.getPrototypeOf) || ([].__proto__ === Array.
  prototype ? function(t) {
    return t.__proto__;
  } : null);
  function vn(t, e) {
    if (t === 1 / 0 || t === -1 / 0 || t !== t || t && t > -1e3 && t < 1e3 || wn.call(/e/, e))
      return e;
    var r = /[0-9](?=(?:[0-9]{3})+(?![0-9]))/g;
    if (typeof t == "number") {
      var n = t < 0 ? -hn(-t) : hn(t);
      if (n !== t) {
        var a = String(n), i = mt.call(e, a.length + 1);
        return B.call(a, r, "$&_") + "." + B.call(B.call(i, /([0-9]{3})/g, "$&_"), /_$/, "");
      }
    }
    return B.call(e, r, "$&_");
  }
  o(vn, "addNumericSeparator");
  var pt = sn(), mn = pt.custom, gn = On(mn) ? mn : null;
  An.exports = /* @__PURE__ */ o(function t(e, r, n, a) {
    var i = r || {};
    if (W(i, "quoteStyle") && i.quoteStyle !== "single" && i.quoteStyle !== "double")
      throw new TypeError('option "quoteStyle" must be "single" or "double"');
    if (W(i, "maxStringLength") && (typeof i.maxStringLength == "number" ? i.maxStringLength < 0 && i.maxStringLength !== 1 / 0 : i.maxStringLength !==
    null))
      throw new TypeError('option "maxStringLength", if provided, must be a positive integer, Infinity, or `null`');
    var s = W(i, "customInspect") ? i.customInspect : !0;
    if (typeof s != "boolean" && s !== "symbol")
      throw new TypeError("option \"customInspect\", if provided, must be `true`, `false`, or `'symbol'`");
    if (W(i, "indent") && i.indent !== null && i.indent !== "	" && !(parseInt(i.indent, 10) === i.indent && i.indent > 0))
      throw new TypeError('option "indent" must be "\\t", an integer > 0, or `null`');
    if (W(i, "numericSeparator") && typeof i.numericSeparator != "boolean")
      throw new TypeError('option "numericSeparator", if provided, must be `true` or `false`');
    var l = i.numericSeparator;
    if (typeof e > "u")
      return "undefined";
    if (e === null)
      return "null";
    if (typeof e == "boolean")
      return e ? "true" : "false";
    if (typeof e == "string")
      return Rn(e, i);
    if (typeof e == "number") {
      if (e === 0)
        return 1 / 0 / e > 0 ? "0" : "-0";
      var u = String(e);
      return l ? vn(e, u) : u;
    }
    if (typeof e == "bigint") {
      var c = String(e) + "n";
      return l ? vn(e, c) : c;
    }
    var f = typeof i.depth > "u" ? 5 : i.depth;
    if (typeof n > "u" && (n = 0), n >= f && f > 0 && typeof e == "object")
      return yt(e) ? "[Array]" : "[Object]";
    var h = li(i, n);
    if (typeof a > "u")
      a = [];
    else if (xn(a, e) >= 0)
      return "[Circular]";
    function p(F, $, ue) {
      if ($ && (a = Vo.call(a), a.push($)), ue) {
        var K = {
          depth: i.depth
        };
        return W(i, "quoteStyle") && (K.quoteStyle = i.quoteStyle), t(F, K, n + 1, a);
      }
      return t(F, i, n + 1, a);
    }
    if (o(p, "inspect"), typeof e == "function" && !bn(e)) {
      var d = Xo(e), y = Te(e, p);
      return "[Function" + (d ? ": " + d : " (anonymous)") + "]" + (y.length > 0 ? " { " + C.call(y, ", ") + " }" : "");
    }
    if (On(e)) {
      var v = ne ? B.call(String(e), /^(Symbol\(.*\))_[^)]*$/, "$1") : ft.call(e);
      return typeof e == "object" && !ne ? he(v) : v;
    }
    if (ai(e)) {
      for (var P = "<" + pn.call(String(e.nodeName)), g = e.attributes || [], E = 0; E < g.length; E++)
        P += " " + g[E].name + "=" + En(jo(g[E].value), "double", i);
      return P += ">", e.childNodes && e.childNodes.length && (P += "..."), P += "</" + pn.call(String(e.nodeName)) + ">", P;
    }
    if (yt(e)) {
      if (e.length === 0)
        return "[]";
      var _ = Te(e, p);
      return h && !ii(_) ? "[" + ht(_, h) + "]" : "[ " + C.call(_, ", ") + " ]";
    }
    if (zo(e)) {
      var D = Te(e, p);
      return !("cause" in Error.prototype) && "cause" in e && !Pn.call(e, "cause") ? "{ [" + String(e) + "] " + C.call(yn.call("[cause]: " +
      p(e.cause), D), ", ") + " }" : D.length === 0 ? "[" + String(e) + "]" : "{ [" + String(e) + "] " + C.call(D, ", ") + " }";
    }
    if (typeof e == "object" && s) {
      if (gn && typeof e[gn] == "function" && pt)
        return pt(e, { depth: f - n });
      if (s !== "symbol" && typeof e.inspect == "function")
        return e.inspect();
    }
    if (Zo(e)) {
      var x = [];
      return un && un.call(e, function(F, $) {
        x.push(p($, e, !0) + " => " + p(F, e));
      }), Sn("Map", Fe.call(e), x, h);
    }
    if (ri(e)) {
      var J = [];
      return cn && cn.call(e, function(F) {
        J.push(p(F, e));
      }), Sn("Set", $e.call(e), J, h);
    }
    if (ei(e))
      return ut("WeakMap");
    if (ni(e))
      return ut("WeakSet");
    if (ti(e))
      return ut("WeakRef");
    if (Qo(e))
      return he(p(Number(e)));
    if (Ko(e))
      return he(p(ct.call(e)));
    if (Jo(e))
      return he(ko.call(e));
    if (Go(e))
      return he(p(String(e)));
    if (typeof window < "u" && e === window)
      return "{ [object Window] }";
    if (e === global)
      return "{ [object globalThis] }";
    if (!Ho(e) && !bn(e)) {
      var m = Te(e, p), w = dn ? dn(e) === Object.prototype : e instanceof Object || e.constructor === Object, O = e instanceof Object ? "" :
      "null prototype", L = !w && M && Object(e) === e && M in e ? mt.call(U(e), 8, -1) : O ? "Object" : "", I = w || typeof e.constructor !=
      "function" ? "" : e.constructor.name ? e.constructor.name + " " : "", R = I + (L || O ? "[" + C.call(yn.call([], L || [], O || []), ":\
 ") + "] " : "");
      return m.length === 0 ? R + "{}" : h ? R + "{" + ht(m, h) + "}" : R + "{ " + C.call(m, ", ") + " }";
    }
    return String(e);
  }, "inspect_");
  function En(t, e, r) {
    var n = (r.quoteStyle || e) === "double" ? '"' : "'";
    return n + t + n;
  }
  o(En, "wrapQuotes");
  function jo(t) {
    return B.call(String(t), /"/g, "&quot;");
  }
  o(jo, "quote");
  function yt(t) {
    return U(t) === "[object Array]" && (!M || !(typeof t == "object" && M in t));
  }
  o(yt, "isArray");
  function Ho(t) {
    return U(t) === "[object Date]" && (!M || !(typeof t == "object" && M in t));
  }
  o(Ho, "isDate");
  function bn(t) {
    return U(t) === "[object RegExp]" && (!M || !(typeof t == "object" && M in t));
  }
  o(bn, "isRegExp");
  function zo(t) {
    return U(t) === "[object Error]" && (!M || !(typeof t == "object" && M in t));
  }
  o(zo, "isError");
  function Go(t) {
    return U(t) === "[object String]" && (!M || !(typeof t == "object" && M in t));
  }
  o(Go, "isString");
  function Qo(t) {
    return U(t) === "[object Number]" && (!M || !(typeof t == "object" && M in t));
  }
  o(Qo, "isNumber");
  function Jo(t) {
    return U(t) === "[object Boolean]" && (!M || !(typeof t == "object" && M in t));
  }
  o(Jo, "isBoolean");
  function On(t) {
    if (ne)
      return t && typeof t == "object" && t instanceof Symbol;
    if (typeof t == "symbol")
      return !0;
    if (!t || typeof t != "object" || !ft)
      return !1;
    try {
      return ft.call(t), !0;
    } catch {
    }
    return !1;
  }
  o(On, "isSymbol");
  function Ko(t) {
    if (!t || typeof t != "object" || !ct)
      return !1;
    try {
      return ct.call(t), !0;
    } catch {
    }
    return !1;
  }
  o(Ko, "isBigInt");
  var Yo = Object.prototype.hasOwnProperty || function(t) {
    return t in this;
  };
  function W(t, e) {
    return Yo.call(t, e);
  }
  o(W, "has");
  function U(t) {
    return qo.call(t);
  }
  o(U, "toStr");
  function Xo(t) {
    if (t.name)
      return t.name;
    var e = Bo.call(Wo.call(t), /^function\s*([\w$]+)/);
    return e ? e[1] : null;
  }
  o(Xo, "nameOf");
  function xn(t, e) {
    if (t.indexOf)
      return t.indexOf(e);
    for (var r = 0, n = t.length; r < n; r++)
      if (t[r] === e)
        return r;
    return -1;
  }
  o(xn, "indexOf");
  function Zo(t) {
    if (!Fe || !t || typeof t != "object")
      return !1;
    try {
      Fe.call(t);
      try {
        $e.call(t);
      } catch {
        return !0;
      }
      return t instanceof Map;
    } catch {
    }
    return !1;
  }
  o(Zo, "isMap");
  function ei(t) {
    if (!de || !t || typeof t != "object")
      return !1;
    try {
      de.call(t, de);
      try {
        ve.call(t, ve);
      } catch {
        return !0;
      }
      return t instanceof WeakMap;
    } catch {
    }
    return !1;
  }
  o(ei, "isWeakMap");
  function ti(t) {
    if (!fn || !t || typeof t != "object")
      return !1;
    try {
      return fn.call(t), !0;
    } catch {
    }
    return !1;
  }
  o(ti, "isWeakRef");
  function ri(t) {
    if (!$e || !t || typeof t != "object")
      return !1;
    try {
      $e.call(t);
      try {
        Fe.call(t);
      } catch {
        return !0;
      }
      return t instanceof Set;
    } catch {
    }
    return !1;
  }
  o(ri, "isSet");
  function ni(t) {
    if (!ve || !t || typeof t != "object")
      return !1;
    try {
      ve.call(t, ve);
      try {
        de.call(t, de);
      } catch {
        return !0;
      }
      return t instanceof WeakSet;
    } catch {
    }
    return !1;
  }
  o(ni, "isWeakSet");
  function ai(t) {
    return !t || typeof t != "object" ? !1 : typeof HTMLElement < "u" && t instanceof HTMLElement ? !0 : typeof t.nodeName == "string" && typeof t.
    getAttribute == "function";
  }
  o(ai, "isElement");
  function Rn(t, e) {
    if (t.length > e.maxStringLength) {
      var r = t.length - e.maxStringLength, n = "... " + r + " more character" + (r > 1 ? "s" : "");
      return Rn(mt.call(t, 0, e.maxStringLength), e) + n;
    }
    var a = B.call(B.call(t, /(['\\])/g, "\\$1"), /[\x00-\x1f]/g, oi);
    return En(a, "single", e);
  }
  o(Rn, "inspectString");
  function oi(t) {
    var e = t.charCodeAt(0), r = {
      8: "b",
      9: "t",
      10: "n",
      12: "f",
      13: "r"
    }[e];
    return r ? "\\" + r : "\\x" + (e < 16 ? "0" : "") + Uo.call(e.toString(16));
  }
  o(oi, "lowbyte");
  function he(t) {
    return "Object(" + t + ")";
  }
  o(he, "markBoxed");
  function ut(t) {
    return t + " { ? }";
  }
  o(ut, "weakCollectionOf");
  function Sn(t, e, r, n) {
    var a = n ? ht(r, n) : C.call(r, ", ");
    return t + " (" + e + ") {" + a + "}";
  }
  o(Sn, "collectionOf");
  function ii(t) {
    for (var e = 0; e < t.length; e++)
      if (xn(t[e], `
`) >= 0)
        return !1;
    return !0;
  }
  o(ii, "singleLineValues");
  function li(t, e) {
    var r;
    if (t.indent === "	")
      r = "	";
    else if (typeof t.indent == "number" && t.indent > 0)
      r = C.call(Array(t.indent + 1), " ");
    else
      return null;
    return {
      base: r,
      prev: C.call(Array(e + 1), r)
    };
  }
  o(li, "getIndent");
  function ht(t, e) {
    if (t.length === 0)
      return "";
    var r = `
` + e.prev + e.base;
    return r + C.call(t, "," + r) + `
` + e.prev;
  }
  o(ht, "indentedJoin");
  function Te(t, e) {
    var r = yt(t), n = [];
    if (r) {
      n.length = t.length;
      for (var a = 0; a < t.length; a++)
        n[a] = W(t, a) ? e(t[a], t) : "";
    }
    var i = typeof st == "function" ? st(t) : [], s;
    if (ne) {
      s = {};
      for (var l = 0; l < i.length; l++)
        s["$" + i[l]] = i[l];
    }
    for (var u in t)
      W(t, u) && (r && String(Number(u)) === u && u < t.length || ne && s["$" + u] instanceof Symbol || (wn.call(/[^\w$]/, u) ? n.push(e(u, t) +
      ": " + e(t[u], t)) : n.push(u + ": " + e(t[u], t))));
    if (typeof st == "function")
      for (var c = 0; c < i.length; c++)
        Pn.call(t, i[c]) && n.push("[" + e(i[c]) + "]: " + e(t[i[c]], t));
    return n;
  }
  o(Te, "arrObjKeys");
});

// ../node_modules/side-channel/index.js
var _n = b((Rs, In) => {
  "use strict";
  var gt = G(), ae = ln(), si = Nn(), ui = gt("%TypeError%"), Ce = gt("%WeakMap%", !0), ke = gt("%Map%", !0), ci = ae("WeakMap.prototype.get",
  !0), fi = ae("WeakMap.prototype.set", !0), pi = ae("WeakMap.prototype.has", !0), yi = ae("Map.prototype.get", !0), hi = ae("Map.prototype.\
set", !0), di = ae("Map.prototype.has", !0), bt = /* @__PURE__ */ o(function(t, e) {
    for (var r = t, n; (n = r.next) !== null; r = n)
      if (n.key === e)
        return r.next = n.next, n.next = t.next, t.next = n, n;
  }, "listGetNode"), vi = /* @__PURE__ */ o(function(t, e) {
    var r = bt(t, e);
    return r && r.value;
  }, "listGet"), mi = /* @__PURE__ */ o(function(t, e, r) {
    var n = bt(t, e);
    n ? n.value = r : t.next = {
      // eslint-disable-line no-param-reassign
      key: e,
      next: t.next,
      value: r
    };
  }, "listSet"), gi = /* @__PURE__ */ o(function(t, e) {
    return !!bt(t, e);
  }, "listHas");
  In.exports = /* @__PURE__ */ o(function() {
    var e, r, n, a = {
      assert: /* @__PURE__ */ o(function(i) {
        if (!a.has(i))
          throw new ui("Side channel does not contain " + si(i));
      }, "assert"),
      get: /* @__PURE__ */ o(function(i) {
        if (Ce && i && (typeof i == "object" || typeof i == "function")) {
          if (e)
            return ci(e, i);
        } else if (ke) {
          if (r)
            return yi(r, i);
        } else if (n)
          return vi(n, i);
      }, "get"),
      has: /* @__PURE__ */ o(function(i) {
        if (Ce && i && (typeof i == "object" || typeof i == "function")) {
          if (e)
            return pi(e, i);
        } else if (ke) {
          if (r)
            return di(r, i);
        } else if (n)
          return gi(n, i);
        return !1;
      }, "has"),
      set: /* @__PURE__ */ o(function(i, s) {
        Ce && i && (typeof i == "object" || typeof i == "function") ? (e || (e = new Ce()), fi(e, i, s)) : ke ? (r || (r = new ke()), hi(r, i,
        s)) : (n || (n = { key: {}, next: null }), mi(n, i, s));
      }, "set")
    };
    return a;
  }, "getSideChannel");
});

// ../node_modules/qs/lib/formats.js
var qe = b((Ns, Dn) => {
  "use strict";
  var bi = String.prototype.replace, Si = /%20/g, St = {
    RFC1738: "RFC1738",
    RFC3986: "RFC3986"
  };
  Dn.exports = {
    default: St.RFC3986,
    formatters: {
      RFC1738: /* @__PURE__ */ o(function(t) {
        return bi.call(t, Si, "+");
      }, "RFC1738"),
      RFC3986: /* @__PURE__ */ o(function(t) {
        return String(t);
      }, "RFC3986")
    },
    RFC1738: St.RFC1738,
    RFC3986: St.RFC3986
  };
});

// ../node_modules/qs/lib/utils.js
var Pt = b((_s, Mn) => {
  "use strict";
  var wi = qe(), wt = Object.prototype.hasOwnProperty, Q = Array.isArray, k = function() {
    for (var t = [], e = 0; e < 256; ++e)
      t.push("%" + ((e < 16 ? "0" : "") + e.toString(16)).toUpperCase());
    return t;
  }(), Pi = /* @__PURE__ */ o(function(e) {
    for (; e.length > 1; ) {
      var r = e.pop(), n = r.obj[r.prop];
      if (Q(n)) {
        for (var a = [], i = 0; i < n.length; ++i)
          typeof n[i] < "u" && a.push(n[i]);
        r.obj[r.prop] = a;
      }
    }
  }, "compactQueue"), Ln = /* @__PURE__ */ o(function(e, r) {
    for (var n = r && r.plainObjects ? /* @__PURE__ */ Object.create(null) : {}, a = 0; a < e.length; ++a)
      typeof e[a] < "u" && (n[a] = e[a]);
    return n;
  }, "arrayToObject"), Ei = /* @__PURE__ */ o(function t(e, r, n) {
    if (!r)
      return e;
    if (typeof r != "object") {
      if (Q(e))
        e.push(r);
      else if (e && typeof e == "object")
        (n && (n.plainObjects || n.allowPrototypes) || !wt.call(Object.prototype, r)) && (e[r] = !0);
      else
        return [e, r];
      return e;
    }
    if (!e || typeof e != "object")
      return [e].concat(r);
    var a = e;
    return Q(e) && !Q(r) && (a = Ln(e, n)), Q(e) && Q(r) ? (r.forEach(function(i, s) {
      if (wt.call(e, s)) {
        var l = e[s];
        l && typeof l == "object" && i && typeof i == "object" ? e[s] = t(l, i, n) : e.push(i);
      } else
        e[s] = i;
    }), e) : Object.keys(r).reduce(function(i, s) {
      var l = r[s];
      return wt.call(i, s) ? i[s] = t(i[s], l, n) : i[s] = l, i;
    }, a);
  }, "merge"), Oi = /* @__PURE__ */ o(function(e, r) {
    return Object.keys(r).reduce(function(n, a) {
      return n[a] = r[a], n;
    }, e);
  }, "assignSingleSource"), xi = /* @__PURE__ */ o(function(t, e, r) {
    var n = t.replace(/\+/g, " ");
    if (r === "iso-8859-1")
      return n.replace(/%[0-9a-f]{2}/gi, unescape);
    try {
      return decodeURIComponent(n);
    } catch {
      return n;
    }
  }, "decode"), Ri = /* @__PURE__ */ o(function(e, r, n, a, i) {
    if (e.length === 0)
      return e;
    var s = e;
    if (typeof e == "symbol" ? s = Symbol.prototype.toString.call(e) : typeof e != "string" && (s = String(e)), n === "iso-8859-1")
      return escape(s).replace(/%u[0-9a-f]{4}/gi, function(f) {
        return "%26%23" + parseInt(f.slice(2), 16) + "%3B";
      });
    for (var l = "", u = 0; u < s.length; ++u) {
      var c = s.charCodeAt(u);
      if (c === 45 || c === 46 || c === 95 || c === 126 || c >= 48 && c <= 57 || c >= 65 && c <= 90 || c >= 97 && c <= 122 || i === wi.RFC1738 &&
      (c === 40 || c === 41)) {
        l += s.charAt(u);
        continue;
      }
      if (c < 128) {
        l = l + k[c];
        continue;
      }
      if (c < 2048) {
        l = l + (k[192 | c >> 6] + k[128 | c & 63]);
        continue;
      }
      if (c < 55296 || c >= 57344) {
        l = l + (k[224 | c >> 12] + k[128 | c >> 6 & 63] + k[128 | c & 63]);
        continue;
      }
      u += 1, c = 65536 + ((c & 1023) << 10 | s.charCodeAt(u) & 1023), l += k[240 | c >> 18] + k[128 | c >> 12 & 63] + k[128 | c >> 6 & 63] +
      k[128 | c & 63];
    }
    return l;
  }, "encode"), Ai = /* @__PURE__ */ o(function(e) {
    for (var r = [{ obj: { o: e }, prop: "o" }], n = [], a = 0; a < r.length; ++a)
      for (var i = r[a], s = i.obj[i.prop], l = Object.keys(s), u = 0; u < l.length; ++u) {
        var c = l[u], f = s[c];
        typeof f == "object" && f !== null && n.indexOf(f) === -1 && (r.push({ obj: s, prop: c }), n.push(f));
      }
    return Pi(r), e;
  }, "compact"), Ni = /* @__PURE__ */ o(function(e) {
    return Object.prototype.toString.call(e) === "[object RegExp]";
  }, "isRegExp"), Ii = /* @__PURE__ */ o(function(e) {
    return !e || typeof e != "object" ? !1 : !!(e.constructor && e.constructor.isBuffer && e.constructor.isBuffer(e));
  }, "isBuffer"), _i = /* @__PURE__ */ o(function(e, r) {
    return [].concat(e, r);
  }, "combine"), Di = /* @__PURE__ */ o(function(e, r) {
    if (Q(e)) {
      for (var n = [], a = 0; a < e.length; a += 1)
        n.push(r(e[a]));
      return n;
    }
    return r(e);
  }, "maybeMap");
  Mn.exports = {
    arrayToObject: Ln,
    assign: Oi,
    combine: _i,
    compact: Ai,
    decode: xi,
    encode: Ri,
    isBuffer: Ii,
    isRegExp: Ni,
    maybeMap: Di,
    merge: Ei
  };
});

// ../node_modules/qs/lib/stringify.js
var qn = b((Ls, kn) => {
  "use strict";
  var $n = _n(), We = Pt(), me = qe(), Li = Object.prototype.hasOwnProperty, Tn = {
    brackets: /* @__PURE__ */ o(function(e) {
      return e + "[]";
    }, "brackets"),
    comma: "comma",
    indices: /* @__PURE__ */ o(function(e, r) {
      return e + "[" + r + "]";
    }, "indices"),
    repeat: /* @__PURE__ */ o(function(e) {
      return e;
    }, "repeat")
  }, q = Array.isArray, Mi = Array.prototype.push, Cn = /* @__PURE__ */ o(function(t, e) {
    Mi.apply(t, q(e) ? e : [e]);
  }, "pushToArray"), Ti = Date.prototype.toISOString, Fn = me.default, T = {
    addQueryPrefix: !1,
    allowDots: !1,
    charset: "utf-8",
    charsetSentinel: !1,
    delimiter: "&",
    encode: !0,
    encoder: We.encode,
    encodeValuesOnly: !1,
    format: Fn,
    formatter: me.formatters[Fn],
    // deprecated
    indices: !1,
    serializeDate: /* @__PURE__ */ o(function(e) {
      return Ti.call(e);
    }, "serializeDate"),
    skipNulls: !1,
    strictNullHandling: !1
  }, Fi = /* @__PURE__ */ o(function(e) {
    return typeof e == "string" || typeof e == "number" || typeof e == "boolean" || typeof e == "symbol" || typeof e == "bigint";
  }, "isNonNullishPrimitive"), Et = {}, $i = /* @__PURE__ */ o(function t(e, r, n, a, i, s, l, u, c, f, h, p, d, y, v, P) {
    for (var g = e, E = P, _ = 0, D = !1; (E = E.get(Et)) !== void 0 && !D; ) {
      var x = E.get(e);
      if (_ += 1, typeof x < "u") {
        if (x === _)
          throw new RangeError("Cyclic object value");
        D = !0;
      }
      typeof E.get(Et) > "u" && (_ = 0);
    }
    if (typeof u == "function" ? g = u(r, g) : g instanceof Date ? g = h(g) : n === "comma" && q(g) && (g = We.maybeMap(g, function(K) {
      return K instanceof Date ? h(K) : K;
    })), g === null) {
      if (i)
        return l && !y ? l(r, T.encoder, v, "key", p) : r;
      g = "";
    }
    if (Fi(g) || We.isBuffer(g)) {
      if (l) {
        var J = y ? r : l(r, T.encoder, v, "key", p);
        return [d(J) + "=" + d(l(g, T.encoder, v, "value", p))];
      }
      return [d(r) + "=" + d(String(g))];
    }
    var m = [];
    if (typeof g > "u")
      return m;
    var w;
    if (n === "comma" && q(g))
      y && l && (g = We.maybeMap(g, l)), w = [{ value: g.length > 0 ? g.join(",") || null : void 0 }];
    else if (q(u))
      w = u;
    else {
      var O = Object.keys(g);
      w = c ? O.sort(c) : O;
    }
    for (var L = a && q(g) && g.length === 1 ? r + "[]" : r, I = 0; I < w.length; ++I) {
      var R = w[I], F = typeof R == "object" && typeof R.value < "u" ? R.value : g[R];
      if (!(s && F === null)) {
        var $ = q(g) ? typeof n == "function" ? n(L, R) : L : L + (f ? "." + R : "[" + R + "]");
        P.set(e, _);
        var ue = $n();
        ue.set(Et, P), Cn(m, t(
          F,
          $,
          n,
          a,
          i,
          s,
          n === "comma" && y && q(g) ? null : l,
          u,
          c,
          f,
          h,
          p,
          d,
          y,
          v,
          ue
        ));
      }
    }
    return m;
  }, "stringify"), Ci = /* @__PURE__ */ o(function(e) {
    if (!e)
      return T;
    if (e.encoder !== null && typeof e.encoder < "u" && typeof e.encoder != "function")
      throw new TypeError("Encoder has to be a function.");
    var r = e.charset || T.charset;
    if (typeof e.charset < "u" && e.charset !== "utf-8" && e.charset !== "iso-8859-1")
      throw new TypeError("The charset option must be either utf-8, iso-8859-1, or undefined");
    var n = me.default;
    if (typeof e.format < "u") {
      if (!Li.call(me.formatters, e.format))
        throw new TypeError("Unknown format option provided.");
      n = e.format;
    }
    var a = me.formatters[n], i = T.filter;
    return (typeof e.filter == "function" || q(e.filter)) && (i = e.filter), {
      addQueryPrefix: typeof e.addQueryPrefix == "boolean" ? e.addQueryPrefix : T.addQueryPrefix,
      allowDots: typeof e.allowDots > "u" ? T.allowDots : !!e.allowDots,
      charset: r,
      charsetSentinel: typeof e.charsetSentinel == "boolean" ? e.charsetSentinel : T.charsetSentinel,
      delimiter: typeof e.delimiter > "u" ? T.delimiter : e.delimiter,
      encode: typeof e.encode == "boolean" ? e.encode : T.encode,
      encoder: typeof e.encoder == "function" ? e.encoder : T.encoder,
      encodeValuesOnly: typeof e.encodeValuesOnly == "boolean" ? e.encodeValuesOnly : T.encodeValuesOnly,
      filter: i,
      format: n,
      formatter: a,
      serializeDate: typeof e.serializeDate == "function" ? e.serializeDate : T.serializeDate,
      skipNulls: typeof e.skipNulls == "boolean" ? e.skipNulls : T.skipNulls,
      sort: typeof e.sort == "function" ? e.sort : null,
      strictNullHandling: typeof e.strictNullHandling == "boolean" ? e.strictNullHandling : T.strictNullHandling
    };
  }, "normalizeStringifyOptions");
  kn.exports = function(t, e) {
    var r = t, n = Ci(e), a, i;
    typeof n.filter == "function" ? (i = n.filter, r = i("", r)) : q(n.filter) && (i = n.filter, a = i);
    var s = [];
    if (typeof r != "object" || r === null)
      return "";
    var l;
    e && e.arrayFormat in Tn ? l = e.arrayFormat : e && "indices" in e ? l = e.indices ? "indices" : "repeat" : l = "indices";
    var u = Tn[l];
    if (e && "commaRoundTrip" in e && typeof e.commaRoundTrip != "boolean")
      throw new TypeError("`commaRoundTrip` must be a boolean, or absent");
    var c = u === "comma" && e && e.commaRoundTrip;
    a || (a = Object.keys(r)), n.sort && a.sort(n.sort);
    for (var f = $n(), h = 0; h < a.length; ++h) {
      var p = a[h];
      n.skipNulls && r[p] === null || Cn(s, $i(
        r[p],
        p,
        u,
        c,
        n.strictNullHandling,
        n.skipNulls,
        n.encode ? n.encoder : null,
        n.filter,
        n.sort,
        n.allowDots,
        n.serializeDate,
        n.format,
        n.formatter,
        n.encodeValuesOnly,
        n.charset,
        f
      ));
    }
    var d = s.join(n.delimiter), y = n.addQueryPrefix === !0 ? "?" : "";
    return n.charsetSentinel && (n.charset === "iso-8859-1" ? y += "utf8=%26%2310003%3B&" : y += "utf8=%E2%9C%93&"), d.length > 0 ? y + d : "";
  };
});

// ../node_modules/qs/lib/parse.js
var Un = b((Ts, Bn) => {
  "use strict";
  var oe = Pt(), Ot = Object.prototype.hasOwnProperty, ki = Array.isArray, N = {
    allowDots: !1,
    allowPrototypes: !1,
    allowSparse: !1,
    arrayLimit: 20,
    charset: "utf-8",
    charsetSentinel: !1,
    comma: !1,
    decoder: oe.decode,
    delimiter: "&",
    depth: 5,
    ignoreQueryPrefix: !1,
    interpretNumericEntities: !1,
    parameterLimit: 1e3,
    parseArrays: !0,
    plainObjects: !1,
    strictNullHandling: !1
  }, qi = /* @__PURE__ */ o(function(t) {
    return t.replace(/&#(\d+);/g, function(e, r) {
      return String.fromCharCode(parseInt(r, 10));
    });
  }, "interpretNumericEntities"), Wn = /* @__PURE__ */ o(function(t, e) {
    return t && typeof t == "string" && e.comma && t.indexOf(",") > -1 ? t.split(",") : t;
  }, "parseArrayValue"), Wi = "utf8=%26%2310003%3B", Bi = "utf8=%E2%9C%93", Ui = /* @__PURE__ */ o(function(e, r) {
    var n = { __proto__: null }, a = r.ignoreQueryPrefix ? e.replace(/^\?/, "") : e, i = r.parameterLimit === 1 / 0 ? void 0 : r.parameterLimit,
    s = a.split(r.delimiter, i), l = -1, u, c = r.charset;
    if (r.charsetSentinel)
      for (u = 0; u < s.length; ++u)
        s[u].indexOf("utf8=") === 0 && (s[u] === Bi ? c = "utf-8" : s[u] === Wi && (c = "iso-8859-1"), l = u, u = s.length);
    for (u = 0; u < s.length; ++u)
      if (u !== l) {
        var f = s[u], h = f.indexOf("]="), p = h === -1 ? f.indexOf("=") : h + 1, d, y;
        p === -1 ? (d = r.decoder(f, N.decoder, c, "key"), y = r.strictNullHandling ? null : "") : (d = r.decoder(f.slice(0, p), N.decoder, c,
        "key"), y = oe.maybeMap(
          Wn(f.slice(p + 1), r),
          function(v) {
            return r.decoder(v, N.decoder, c, "value");
          }
        )), y && r.interpretNumericEntities && c === "iso-8859-1" && (y = qi(y)), f.indexOf("[]=") > -1 && (y = ki(y) ? [y] : y), Ot.call(n,
        d) ? n[d] = oe.combine(n[d], y) : n[d] = y;
      }
    return n;
  }, "parseQueryStringValues"), Vi = /* @__PURE__ */ o(function(t, e, r, n) {
    for (var a = n ? e : Wn(e, r), i = t.length - 1; i >= 0; --i) {
      var s, l = t[i];
      if (l === "[]" && r.parseArrays)
        s = [].concat(a);
      else {
        s = r.plainObjects ? /* @__PURE__ */ Object.create(null) : {};
        var u = l.charAt(0) === "[" && l.charAt(l.length - 1) === "]" ? l.slice(1, -1) : l, c = parseInt(u, 10);
        !r.parseArrays && u === "" ? s = { 0: a } : !isNaN(c) && l !== u && String(c) === u && c >= 0 && r.parseArrays && c <= r.arrayLimit ?
        (s = [], s[c] = a) : u !== "__proto__" && (s[u] = a);
      }
      a = s;
    }
    return a;
  }, "parseObject"), ji = /* @__PURE__ */ o(function(e, r, n, a) {
    if (e) {
      var i = n.allowDots ? e.replace(/\.([^.[]+)/g, "[$1]") : e, s = /(\[[^[\]]*])/, l = /(\[[^[\]]*])/g, u = n.depth > 0 && s.exec(i), c = u ?
      i.slice(0, u.index) : i, f = [];
      if (c) {
        if (!n.plainObjects && Ot.call(Object.prototype, c) && !n.allowPrototypes)
          return;
        f.push(c);
      }
      for (var h = 0; n.depth > 0 && (u = l.exec(i)) !== null && h < n.depth; ) {
        if (h += 1, !n.plainObjects && Ot.call(Object.prototype, u[1].slice(1, -1)) && !n.allowPrototypes)
          return;
        f.push(u[1]);
      }
      return u && f.push("[" + i.slice(u.index) + "]"), Vi(f, r, n, a);
    }
  }, "parseQueryStringKeys"), Hi = /* @__PURE__ */ o(function(e) {
    if (!e)
      return N;
    if (e.decoder !== null && e.decoder !== void 0 && typeof e.decoder != "function")
      throw new TypeError("Decoder has to be a function.");
    if (typeof e.charset < "u" && e.charset !== "utf-8" && e.charset !== "iso-8859-1")
      throw new TypeError("The charset option must be either utf-8, iso-8859-1, or undefined");
    var r = typeof e.charset > "u" ? N.charset : e.charset;
    return {
      allowDots: typeof e.allowDots > "u" ? N.allowDots : !!e.allowDots,
      allowPrototypes: typeof e.allowPrototypes == "boolean" ? e.allowPrototypes : N.allowPrototypes,
      allowSparse: typeof e.allowSparse == "boolean" ? e.allowSparse : N.allowSparse,
      arrayLimit: typeof e.arrayLimit == "number" ? e.arrayLimit : N.arrayLimit,
      charset: r,
      charsetSentinel: typeof e.charsetSentinel == "boolean" ? e.charsetSentinel : N.charsetSentinel,
      comma: typeof e.comma == "boolean" ? e.comma : N.comma,
      decoder: typeof e.decoder == "function" ? e.decoder : N.decoder,
      delimiter: typeof e.delimiter == "string" || oe.isRegExp(e.delimiter) ? e.delimiter : N.delimiter,
      // eslint-disable-next-line no-implicit-coercion, no-extra-parens
      depth: typeof e.depth == "number" || e.depth === !1 ? +e.depth : N.depth,
      ignoreQueryPrefix: e.ignoreQueryPrefix === !0,
      interpretNumericEntities: typeof e.interpretNumericEntities == "boolean" ? e.interpretNumericEntities : N.interpretNumericEntities,
      parameterLimit: typeof e.parameterLimit == "number" ? e.parameterLimit : N.parameterLimit,
      parseArrays: e.parseArrays !== !1,
      plainObjects: typeof e.plainObjects == "boolean" ? e.plainObjects : N.plainObjects,
      strictNullHandling: typeof e.strictNullHandling == "boolean" ? e.strictNullHandling : N.strictNullHandling
    };
  }, "normalizeParseOptions");
  Bn.exports = function(t, e) {
    var r = Hi(e);
    if (t === "" || t === null || typeof t > "u")
      return r.plainObjects ? /* @__PURE__ */ Object.create(null) : {};
    for (var n = typeof t == "string" ? Ui(t, r) : t, a = r.plainObjects ? /* @__PURE__ */ Object.create(null) : {}, i = Object.keys(n), s = 0; s <
    i.length; ++s) {
      var l = i[s], u = ji(l, n[l], r, typeof t == "string");
      a = oe.merge(a, u, r);
    }
    return r.allowSparse === !0 ? a : oe.compact(a);
  };
});

// ../node_modules/qs/lib/index.js
var jn = b(($s, Vn) => {
  "use strict";
  var zi = qn(), Gi = Un(), Qi = qe();
  Vn.exports = {
    formats: Qi,
    parse: Gi,
    stringify: zi
  };
});

// src/router/utils.ts
import { once as Ji } from "@storybook/core/client-logger";

// ../node_modules/dequal/dist/index.mjs
var Mt = Object.prototype.hasOwnProperty;
function Tt(t, e, r) {
  for (r of t.keys())
    if (Y(r, e)) return r;
}
o(Tt, "find");
function Y(t, e) {
  var r, n, a;
  if (t === e) return !0;
  if (t && e && (r = t.constructor) === e.constructor) {
    if (r === Date) return t.getTime() === e.getTime();
    if (r === RegExp) return t.toString() === e.toString();
    if (r === Array) {
      if ((n = t.length) === e.length)
        for (; n-- && Y(t[n], e[n]); ) ;
      return n === -1;
    }
    if (r === Set) {
      if (t.size !== e.size)
        return !1;
      for (n of t)
        if (a = n, a && typeof a == "object" && (a = Tt(e, a), !a) || !e.has(a)) return !1;
      return !0;
    }
    if (r === Map) {
      if (t.size !== e.size)
        return !1;
      for (n of t)
        if (a = n[0], a && typeof a == "object" && (a = Tt(e, a), !a) || !Y(n[1], e.get(a)))
          return !1;
      return !0;
    }
    if (r === ArrayBuffer)
      t = new Uint8Array(t), e = new Uint8Array(e);
    else if (r === DataView) {
      if ((n = t.byteLength) === e.byteLength)
        for (; n-- && t.getInt8(n) === e.getInt8(n); ) ;
      return n === -1;
    }
    if (ArrayBuffer.isView(t)) {
      if ((n = t.byteLength) === e.byteLength)
        for (; n-- && t[n] === e[n]; ) ;
      return n === -1;
    }
    if (!r || typeof t == "object") {
      n = 0;
      for (r in t)
        if (Mt.call(t, r) && ++n && !Mt.call(e, r) || !(r in e) || !Y(t[r], e[r])) return !1;
      return Object.keys(e).length === n;
    }
  }
  return t !== t && e !== e;
}
o(Y, "dequal");

// src/router/utils.ts
var ge = Ke(ar(), 1), Ue = Ke(ir(), 1), Ve = Ke(jn(), 1);

// ../node_modules/ts-dedent/esm/index.js
function Hn(t) {
  for (var e = [], r = 1; r < arguments.length; r++)
    e[r - 1] = arguments[r];
  var n = Array.from(typeof t == "string" ? [t] : t);
  n[n.length - 1] = n[n.length - 1].replace(/\r?\n([\t ]*)$/, "");
  var a = n.reduce(function(l, u) {
    var c = u.match(/\n([\t ]+|(?!\s).)/g);
    return c ? l.concat(c.map(function(f) {
      var h, p;
      return (p = (h = f.match(/[\t ]/g)) === null || h === void 0 ? void 0 : h.length) !== null && p !== void 0 ? p : 0;
    })) : l;
  }, []);
  if (a.length) {
    var i = new RegExp(`
[	 ]{` + Math.min.apply(Math, a) + "}", "g");
    n = n.map(function(l) {
      return l.replace(i, `
`);
    });
  }
  n[0] = n[0].replace(/^\r?\n/, "");
  var s = n[0];
  return e.forEach(function(l, u) {
    var c = s.match(/(?:^|\n)( *)$/), f = c ? c[1] : "", h = l;
    typeof l == "string" && l.includes(`
`) && (h = String(l).split(`
`).map(function(p, d) {
      return d === 0 ? p : "" + f + p;
    }).join(`
`)), s += h + n[u + 1];
  }), s;
}
o(Hn, "dedent");

// src/router/utils.ts
var Ki = /\/([^/]+)\/(?:(.*)_)?([^/]+)?/, Gn = (0, Ue.default)(1e3)((t) => {
  let e = {
    viewMode: void 0,
    storyId: void 0,
    refId: void 0
  };
  if (t) {
    let [, r, n, a] = t.toLowerCase().match(Ki) || [];
    r && Object.assign(e, {
      viewMode: r,
      storyId: a,
      refId: n
    });
  }
  return e;
}), Be = Symbol("Deeply equal"), xt = /* @__PURE__ */ o((t, e) => {
  if (typeof t != typeof e) return e;
  if (Y(t, e)) return Be;
  if (Array.isArray(t) && Array.isArray(e)) {
    let r = e.reduce((n, a, i) => {
      let s = xt(t[i], a);
      return s !== Be && (n[i] = s), n;
    }, new Array(e.length));
    return e.length >= t.length ? r : r.concat(new Array(t.length - e.length).fill(void 0));
  }
  return (0, ge.default)(t) && (0, ge.default)(e) ? Object.keys({ ...t, ...e }).reduce((r, n) => {
    let a = xt(t?.[n], e?.[n]);
    return a === Be ? r : Object.assign(r, { [n]: a });
  }, {}) : e;
}, "deepDiff"), zn = /^[a-zA-Z0-9 _-]*$/, Yi = /^-?[0-9]+(\.[0-9]+)?$/, Qn = /^#([a-f0-9]{3,4}|[a-f0-9]{6}|[a-f0-9]{8})$/i, Jn = /^(rgba?|hsla?)\(([0-9]{1,3}),\s?([0-9]{1,3})%?,\s?([0-9]{1,3})%?,?\s?([0-9](\.[0-9]{1,2})?)?\)$/i,
Rt = /* @__PURE__ */ o((t = "", e) => t === null || t === "" || !zn.test(t) ? !1 : e == null || e instanceof Date || typeof e == "number" ||
typeof e == "boolean" ? !0 : typeof e == "string" ? zn.test(e) || Yi.test(e) || Qn.test(e) || Jn.test(e) : Array.isArray(e) ? e.every((r) => Rt(
t, r)) : (0, ge.default)(e) ? Object.entries(e).every(([r, n]) => Rt(r, n)) : !1, "validateArgs"), At = /* @__PURE__ */ o((t) => t === void 0 ?
"!undefined" : t === null ? "!null" : typeof t == "string" ? Qn.test(t) ? `!hex(${t.slice(1)})` : Jn.test(t) ? `!${t.replace(/[\s%]/g, "")}` :
t : typeof t == "boolean" ? `!${t}` : Array.isArray(t) ? t.map(At) : (0, ge.default)(t) ? Object.entries(t).reduce(
  (e, [r, n]) => Object.assign(e, { [r]: At(n) }),
  {}
) : t, "encodeSpecialValues"), Xi = {
  encode: !1,
  // we handle URL encoding ourselves
  delimiter: ";",
  // we don't actually create multiple query params
  allowDots: !0,
  // encode objects using dot notation: obj.key=val
  format: "RFC1738",
  // encode spaces using the + sign
  serializeDate: /* @__PURE__ */ o((t) => `!date(${t.toISOString()})`, "serializeDate")
}, Us = /* @__PURE__ */ o((t, e) => {
  let r = xt(t, e);
  if (!r || r === Be) return "";
  let n = Object.entries(r).reduce((a, [i, s]) => Rt(i, s) ? Object.assign(a, { [i]: s }) : (Ji.warn(Hn`
      Omitted potentially unsafe URL args.

      More info: https://storybook.js.org/docs/react/writing-stories/args#setting-args-through-the-url
    `), a), {});
  return Ve.default.stringify(At(n), Xi).replace(/ /g, "+").split(";").map((a) => a.replace("=", ":")).join(";");
}, "buildArgsParam"), Nt = (0, Ue.default)(1e3)(
  (t) => t !== void 0 ? Ve.default.parse(t, { ignoreQueryPrefix: !0 }) : {}
), Vs = /* @__PURE__ */ o((t) => Nt(t.search), "queryFromLocation"), js = /* @__PURE__ */ o((t) => Ve.default.stringify(t, { addQueryPrefix: !0,
encode: !1 }), "stringifyQuery"), Kn = (0, Ue.default)(1e3)((t, e, r = !0) => {
  if (r) {
    if (typeof e != "string")
      throw new Error("startsWith only works with string targets");
    return t && t.startsWith(e) ? { path: t } : null;
  }
  let n = typeof e == "string" && t === e, a = t && e && t.match(e);
  return n || a ? { path: t } : null;
});

// ../node_modules/@storybook/global/dist/index.mjs
var Yn = (() => {
  let t;
  return typeof window < "u" ? t = window : typeof globalThis < "u" ? t = globalThis : typeof global < "u" ? t = global : typeof self < "u" ?
  t = self : t = {}, t;
})();

// src/router/router.tsx
import xe, { useCallback as wl } from "react";

// ../node_modules/react-router-dom/index.js
import { useRef as fl, useState as pl, useLayoutEffect as yl, createElement as Lt, forwardRef as sa, useCallback as hl, useMemo as Ru } from "react";

// ../node_modules/@babel/runtime/helpers/esm/extends.js
function ie() {
  return ie = Object.assign ? Object.assign.bind() : function(t) {
    for (var e = 1; e < arguments.length; e++) {
      var r = arguments[e];
      for (var n in r)
        Object.prototype.hasOwnProperty.call(r, n) && (t[n] = r[n]);
    }
    return t;
  }, ie.apply(this, arguments);
}
o(ie, "_extends");

// ../node_modules/history/index.js
var V;
(function(t) {
  t.Pop = "POP", t.Push = "PUSH", t.Replace = "REPLACE";
})(V || (V = {}));
var Xn = function(t) {
  return Object.freeze(t);
};
function Zi(t, e) {
  if (!t) {
    typeof console < "u" && console.warn(e);
    try {
      throw new Error(e);
    } catch {
    }
  }
}
o(Zi, "warning");
var Zn = "beforeunload";
var el = "popstate";
function ra(t) {
  t === void 0 && (t = {});
  var e = t, r = e.window, n = r === void 0 ? document.defaultView : r, a = n.history;
  function i() {
    var m = n.location, w = m.pathname, O = m.search, L = m.hash, I = a.state || {};
    return [I.idx, Xn({
      pathname: w,
      search: O,
      hash: L,
      state: I.usr || null,
      key: I.key || "default"
    })];
  }
  o(i, "getIndexAndLocation");
  var s = null;
  function l() {
    if (s)
      d.call(s), s = null;
    else {
      var m = V.Pop, w = i(), O = w[0], L = w[1];
      if (d.length)
        if (O != null) {
          var I = f - O;
          I && (s = {
            action: m,
            location: L,
            retry: /* @__PURE__ */ o(function() {
              x(I * -1);
            }, "retry")
          }, x(I));
        } else
          Zi(
            !1,
            // TODO: Write up a doc that explains our blocking strategy in
            // detail and link to it here so people can understand better what
            // is going on and how to avoid it.
            "You are trying to block a POP navigation to a location that was not created by the history library. The block will fail silentl\
y in production, but in general you should do all navigation with the history library (instead of using window.history.pushState directly) t\
o avoid this situation."
          );
      else
        E(m);
    }
  }
  o(l, "handlePop"), n.addEventListener(el, l);
  var u = V.Pop, c = i(), f = c[0], h = c[1], p = ta(), d = ta();
  f == null && (f = 0, a.replaceState(ie({}, a.state, {
    idx: f
  }), ""));
  function y(m) {
    return typeof m == "string" ? m : je(m);
  }
  o(y, "createHref");
  function v(m, w) {
    return w === void 0 && (w = null), Xn(ie({
      pathname: h.pathname,
      hash: "",
      search: ""
    }, typeof m == "string" ? le(m) : m, {
      state: w,
      key: tl()
    }));
  }
  o(v, "getNextLocation");
  function P(m, w) {
    return [{
      usr: m.state,
      key: m.key,
      idx: w
    }, y(m)];
  }
  o(P, "getHistoryStateAndUrl");
  function g(m, w, O) {
    return !d.length || (d.call({
      action: m,
      location: w,
      retry: O
    }), !1);
  }
  o(g, "allowTx");
  function E(m) {
    u = m;
    var w = i();
    f = w[0], h = w[1], p.call({
      action: u,
      location: h
    });
  }
  o(E, "applyTx");
  function _(m, w) {
    var O = V.Push, L = v(m, w);
    function I() {
      _(m, w);
    }
    if (o(I, "retry"), g(O, L, I)) {
      var R = P(L, f + 1), F = R[0], $ = R[1];
      try {
        a.pushState(F, "", $);
      } catch {
        n.location.assign($);
      }
      E(O);
    }
  }
  o(_, "push");
  function D(m, w) {
    var O = V.Replace, L = v(m, w);
    function I() {
      D(m, w);
    }
    if (o(I, "retry"), g(O, L, I)) {
      var R = P(L, f), F = R[0], $ = R[1];
      a.replaceState(F, "", $), E(O);
    }
  }
  o(D, "replace");
  function x(m) {
    a.go(m);
  }
  o(x, "go");
  var J = {
    get action() {
      return u;
    },
    get location() {
      return h;
    },
    createHref: y,
    push: _,
    replace: D,
    go: x,
    back: /* @__PURE__ */ o(function() {
      x(-1);
    }, "back"),
    forward: /* @__PURE__ */ o(function() {
      x(1);
    }, "forward"),
    listen: /* @__PURE__ */ o(function(w) {
      return p.push(w);
    }, "listen"),
    block: /* @__PURE__ */ o(function(w) {
      var O = d.push(w);
      return d.length === 1 && n.addEventListener(Zn, ea), function() {
        O(), d.length || n.removeEventListener(Zn, ea);
      };
    }, "block")
  };
  return J;
}
o(ra, "createBrowserHistory");
function ea(t) {
  t.preventDefault(), t.returnValue = "";
}
o(ea, "promptBeforeUnload");
function ta() {
  var t = [];
  return {
    get length() {
      return t.length;
    },
    push: /* @__PURE__ */ o(function(r) {
      return t.push(r), function() {
        t = t.filter(function(n) {
          return n !== r;
        });
      };
    }, "push"),
    call: /* @__PURE__ */ o(function(r) {
      t.forEach(function(n) {
        return n && n(r);
      });
    }, "call")
  };
}
o(ta, "createEvents");
function tl() {
  return Math.random().toString(36).substr(2, 8);
}
o(tl, "createKey");
function je(t) {
  var e = t.pathname, r = e === void 0 ? "/" : e, n = t.search, a = n === void 0 ? "" : n, i = t.hash, s = i === void 0 ? "" : i;
  return a && a !== "?" && (r += a.charAt(0) === "?" ? a : "?" + a), s && s !== "#" && (r += s.charAt(0) === "#" ? s : "#" + s), r;
}
o(je, "createPath");
function le(t) {
  var e = {};
  if (t) {
    var r = t.indexOf("#");
    r >= 0 && (e.hash = t.substr(r), t = t.substr(0, r));
    var n = t.indexOf("?");
    n >= 0 && (e.search = t.substr(n), t = t.substr(0, n)), t && (e.pathname = t);
  }
  return e;
}
o(le, "parsePath");

// ../node_modules/react-router/index.js
import { createContext as _t, useRef as rl, useState as tu, useLayoutEffect as ru, createElement as na, useContext as se, useEffect as nl, useMemo as It,
useCallback as al, Children as nu, isValidElement as au, Fragment as ou } from "react";
function He(t, e) {
  if (!t) throw new Error(e);
}
o(He, "invariant");
function aa(t, e) {
  if (!t) {
    typeof console < "u" && console.warn(e);
    try {
      throw new Error(e);
    } catch {
    }
  }
}
o(aa, "warning");
var be = /* @__PURE__ */ _t(null);
be.displayName = "Navigation";
var Se = /* @__PURE__ */ _t(null);
Se.displayName = "Location";
var ze = /* @__PURE__ */ _t({
  outlet: null,
  matches: []
});
ze.displayName = "Route";
function we(t) {
  let {
    basename: e = "/",
    children: r = null,
    location: n,
    navigationType: a = V.Pop,
    navigator: i,
    static: s = !1
  } = t;
  Pe() && He(!1, "You cannot render a <Router> inside another <Router>. You should never have more than one in your app.");
  let l = sl(e), u = It(() => ({
    basename: l,
    navigator: i,
    static: s
  }), [l, i, s]);
  typeof n == "string" && (n = le(n));
  let {
    pathname: c = "/",
    search: f = "",
    hash: h = "",
    state: p = null,
    key: d = "default"
  } = n, y = It(() => {
    let v = ll(c, l);
    return v == null ? null : {
      pathname: v,
      search: f,
      hash: h,
      state: p,
      key: d
    };
  }, [l, c, f, h, p, d]);
  return aa(y != null, '<Router basename="' + l + '"> is not able to match the URL ' + ('"' + c + f + h + '" because it does not start with \
the ') + "basename, so the <Router> won't render anything."), y == null ? null : /* @__PURE__ */ na(be.Provider, {
    value: u
  }, /* @__PURE__ */ na(Se.Provider, {
    children: r,
    value: {
      location: y,
      navigationType: a
    }
  }));
}
o(we, "Router");
function Dt(t) {
  Pe() || He(
    !1,
    // TODO: This error is probably because they somehow have 2 versions of the
    // router loaded. We can help them understand how to avoid that.
    "useHref() may be used only in the context of a <Router> component."
  );
  let {
    basename: e,
    navigator: r
  } = se(be), {
    hash: n,
    pathname: a,
    search: i
  } = Oe(t), s = a;
  if (e !== "/") {
    let l = il(t), u = l != null && l.endsWith("/");
    s = a === "/" ? e + (u ? "/" : "") : la([e, a]);
  }
  return r.createHref({
    pathname: s,
    search: i,
    hash: n
  });
}
o(Dt, "useHref");
function Pe() {
  return se(Se) != null;
}
o(Pe, "useInRouterContext");
function j() {
  return Pe() || He(
    !1,
    // TODO: This error is probably because they somehow have 2 versions of the
    // router loaded. We can help them understand how to avoid that.
    "useLocation() may be used only in the context of a <Router> component."
  ), se(Se).location;
}
o(j, "useLocation");
function Ee() {
  Pe() || He(
    !1,
    // TODO: This error is probably because they somehow have 2 versions of the
    // router loaded. We can help them understand how to avoid that.
    "useNavigate() may be used only in the context of a <Router> component."
  );
  let {
    basename: t,
    navigator: e
  } = se(be), {
    matches: r
  } = se(ze), {
    pathname: n
  } = j(), a = JSON.stringify(r.map((l) => l.pathnameBase)), i = rl(!1);
  return nl(() => {
    i.current = !0;
  }), al(function(l, u) {
    if (u === void 0 && (u = {}), aa(i.current, "You should call navigate() in a React.useEffect(), not when your component is first rendere\
d."), !i.current) return;
    if (typeof l == "number") {
      e.go(l);
      return;
    }
    let c = ia(l, JSON.parse(a), n);
    t !== "/" && (c.pathname = la([t, c.pathname])), (u.replace ? e.replace : e.push)(c, u.state);
  }, [t, e, a, n]);
}
o(Ee, "useNavigate");
function Oe(t) {
  let {
    matches: e
  } = se(ze), {
    pathname: r
  } = j(), n = JSON.stringify(e.map((a) => a.pathnameBase));
  return It(() => ia(t, JSON.parse(n), r), [t, n, r]);
}
o(Oe, "useResolvedPath");
function oa(t, e) {
  e === void 0 && (e = "/");
  let {
    pathname: r,
    search: n = "",
    hash: a = ""
  } = typeof t == "string" ? le(t) : t;
  return {
    pathname: r ? r.startsWith("/") ? r : ol(r, e) : e,
    search: ul(n),
    hash: cl(a)
  };
}
o(oa, "resolvePath");
function ol(t, e) {
  let r = e.replace(/\/+$/, "").split("/");
  return t.split("/").forEach((a) => {
    a === ".." ? r.length > 1 && r.pop() : a !== "." && r.push(a);
  }), r.length > 1 ? r.join("/") : "/";
}
o(ol, "resolvePathname");
function ia(t, e, r) {
  let n = typeof t == "string" ? le(t) : t, a = t === "" || n.pathname === "" ? "/" : n.pathname, i;
  if (a == null)
    i = r;
  else {
    let l = e.length - 1;
    if (a.startsWith("..")) {
      let u = a.split("/");
      for (; u[0] === ".."; )
        u.shift(), l -= 1;
      n.pathname = u.join("/");
    }
    i = l >= 0 ? e[l] : "/";
  }
  let s = oa(n, i);
  return a && a !== "/" && a.endsWith("/") && !s.pathname.endsWith("/") && (s.pathname += "/"), s;
}
o(ia, "resolveTo");
function il(t) {
  return t === "" || t.pathname === "" ? "/" : typeof t == "string" ? le(t).pathname : t.pathname;
}
o(il, "getToPathname");
function ll(t, e) {
  if (e === "/") return t;
  if (!t.toLowerCase().startsWith(e.toLowerCase()))
    return null;
  let r = t.charAt(e.length);
  return r && r !== "/" ? null : t.slice(e.length) || "/";
}
o(ll, "stripBasename");
var la = /* @__PURE__ */ o((t) => t.join("/").replace(/\/\/+/g, "/"), "joinPaths"), sl = /* @__PURE__ */ o((t) => t.replace(/\/+$/, "").replace(
/^\/*/, "/"), "normalizePathname"), ul = /* @__PURE__ */ o((t) => !t || t === "?" ? "" : t.startsWith("?") ? t : "?" + t, "normalizeSearch"),
cl = /* @__PURE__ */ o((t) => !t || t === "#" ? "" : t.startsWith("#") ? t : "#" + t, "normalizeHash");

// ../node_modules/react-router-dom/index.js
function Ge() {
  return Ge = Object.assign || function(t) {
    for (var e = 1; e < arguments.length; e++) {
      var r = arguments[e];
      for (var n in r)
        Object.prototype.hasOwnProperty.call(r, n) && (t[n] = r[n]);
    }
    return t;
  }, Ge.apply(this, arguments);
}
o(Ge, "_extends");
function ua(t, e) {
  if (t == null) return {};
  var r = {}, n = Object.keys(t), a, i;
  for (i = 0; i < n.length; i++)
    a = n[i], !(e.indexOf(a) >= 0) && (r[a] = t[a]);
  return r;
}
o(ua, "_objectWithoutPropertiesLoose");
var dl = ["onClick", "reloadDocument", "replace", "state", "target", "to"], vl = ["aria-current", "caseSensitive", "className", "end", "styl\
e", "to"];
function ca(t) {
  let {
    basename: e,
    children: r,
    window: n
  } = t, a = fl();
  a.current == null && (a.current = ra({
    window: n
  }));
  let i = a.current, [s, l] = pl({
    action: i.action,
    location: i.location
  });
  return yl(() => i.listen(l), [i]), /* @__PURE__ */ Lt(we, {
    basename: e,
    children: r,
    location: s.location,
    navigationType: s.action,
    navigator: i
  });
}
o(ca, "BrowserRouter");
function ml(t) {
  return !!(t.metaKey || t.altKey || t.ctrlKey || t.shiftKey);
}
o(ml, "isModifiedEvent");
var Qe = /* @__PURE__ */ sa(/* @__PURE__ */ o(function(e, r) {
  let {
    onClick: n,
    reloadDocument: a,
    replace: i = !1,
    state: s,
    target: l,
    to: u
  } = e, c = ua(e, dl), f = Dt(u), h = bl(u, {
    replace: i,
    state: s,
    target: l
  });
  function p(d) {
    n && n(d), !d.defaultPrevented && !a && h(d);
  }
  return o(p, "handleClick"), // eslint-disable-next-line jsx-a11y/anchor-has-content
  /* @__PURE__ */ Lt("a", Ge({}, c, {
    href: f,
    onClick: p,
    ref: r,
    target: l
  }));
}, "LinkWithRef"));
Qe.displayName = "Link";
var gl = /* @__PURE__ */ sa(/* @__PURE__ */ o(function(e, r) {
  let {
    "aria-current": n = "page",
    caseSensitive: a = !1,
    className: i = "",
    end: s = !1,
    style: l,
    to: u
  } = e, c = ua(e, vl), f = j(), h = Oe(u), p = f.pathname, d = h.pathname;
  a || (p = p.toLowerCase(), d = d.toLowerCase());
  let y = p === d || !s && p.startsWith(d) && p.charAt(d.length) === "/", v = y ? n : void 0, P;
  typeof i == "function" ? P = i({
    isActive: y
  }) : P = [i, y ? "active" : null].filter(Boolean).join(" ");
  let g = typeof l == "function" ? l({
    isActive: y
  }) : l;
  return /* @__PURE__ */ Lt(Qe, Ge({}, c, {
    "aria-current": v,
    className: P,
    ref: r,
    style: g,
    to: u
  }));
}, "NavLinkWithRef"));
gl.displayName = "NavLink";
function bl(t, e) {
  let {
    target: r,
    replace: n,
    state: a
  } = e === void 0 ? {} : e, i = Ee(), s = j(), l = Oe(t);
  return hl((u) => {
    if (u.button === 0 && // Ignore everything but left clicks
    (!r || r === "_self") && // Let browser handle "target=_blank" etc.
    !ml(u)) {
      u.preventDefault();
      let c = !!n || je(s) === je(l);
      i(t, {
        replace: c,
        state: a
      });
    }
  }, [s, i, l, n, a, r, t]);
}
o(bl, "useLinkClickHandler");

// src/router/router.tsx
var { document: fa } = Yn, Pl = /* @__PURE__ */ o(() => `${fa.location.pathname}?`, "getBase"), Tu = /* @__PURE__ */ o(() => {
  let t = Ee();
  return wl((e, { plain: r, ...n } = {}) => {
    if (typeof e == "string" && e.startsWith("#")) {
      fa.location.hash = e;
      return;
    }
    if (typeof e == "string") {
      let a = r ? e : `?path=${e}`;
      return t(a, n);
    }
    if (typeof e == "number")
      return t(e);
  }, []);
}, "useNavigate"), El = /* @__PURE__ */ o(({ to: t, children: e, ...r }) => /* @__PURE__ */ xe.createElement(Qe, { to: `${Pl()}path=${t}`, ...r },
e), "Link");
El.displayName = "QueryLink";
var pa = /* @__PURE__ */ o(({ children: t }) => {
  let e = j(), { path: r, singleStory: n } = Nt(e.search), { viewMode: a, storyId: i, refId: s } = Gn(r);
  return /* @__PURE__ */ xe.createElement(xe.Fragment, null, t({
    path: r || "/",
    location: e,
    viewMode: a,
    storyId: i,
    refId: s,
    singleStory: n === "true"
  }));
}, "Location");
pa.displayName = "QueryLocation";
function ya({
  children: t,
  path: e,
  startsWith: r = !1
}) {
  return /* @__PURE__ */ xe.createElement(pa, null, ({ path: n, ...a }) => t({
    match: Kn(n, e, r),
    ...a
  }));
}
o(ya, "Match");
ya.displayName = "QueryMatch";
function Ol(t) {
  let { children: e, ...r } = t;
  return r.startsWith === void 0 && (r.startsWith = !1), /* @__PURE__ */ xe.createElement(ya, { ...r }, ({ match: a }) => a ? e : null);
}
o(Ol, "Route");
Ol.displayName = "Route";
var Fu = /* @__PURE__ */ o((...t) => ca(...t), "LocationProvider"), $u = /* @__PURE__ */ o((...t) => we(...t), "BaseLocationProvider");
export {
  $u as BaseLocationProvider,
  Be as DEEPLY_EQUAL,
  El as Link,
  pa as Location,
  Fu as LocationProvider,
  ya as Match,
  Ol as Route,
  Us as buildArgsParam,
  xt as deepDiff,
  Kn as getMatch,
  Gn as parsePath,
  Vs as queryFromLocation,
  Nt as queryFromString,
  js as stringifyQuery,
  Tu as useNavigate
};
