// victory-selection-container/es/victory-selection-container.js
import React from "react";
import { Rect, VictoryContainer } from "victory-core";
import { SelectionHelpers } from "./selection-helpers";
export function selectionContainerMixin(Base) {
  // @ts-expect-error "TS2545: A mixin class must have a constructor with a single rest parameter of type 'any[]'."
  return class VictorySelectionContainer extends Base {
    static displayName = "VictorySelectionContainer";
    static defaultProps = {
      ...VictoryContainer.defaultProps,
      activateSelectedData: true,
      allowSelection: true,
      selectionComponent: /*#__PURE__*/React.createElement(Rect, null),
      selectionStyle: {
        stroke: "transparent",
        fill: "black",
        fillOpacity: 0.1
      }
    };
    static defaultEvents(props) {
      return [{
        target: "parent",
        eventHandlers: {
          onMouseDown: (evt, targetProps) => {
            return props.disable ? {} : SelectionHelpers.onMouseDown(evt, targetProps);
          },
          onTouchStart: (evt, targetProps) => {
            return props.disable ? {} : SelectionHelpers.onMouseDown(evt, targetProps);
          },
          onMouseMove: (evt, targetProps) => {
            return props.disable ? {} : SelectionHelpers.onMouseMove(evt, targetProps);
          },
          onTouchMove: (evt, targetProps) => {
            return props.disable ? {} : SelectionHelpers.onMouseMove(evt, targetProps);
          },
          onMouseUp: (evt, targetProps) => {
            return props.disable ? {} : SelectionHelpers.onMouseUp(evt, targetProps);
          },
          onTouchEnd: (evt, targetProps) => {
            return props.disable ? {} : SelectionHelpers.onMouseUp(evt, targetProps);
          }
        }
      }];
    }
    getRect(props) {
      const {
        x1,
        x2,
        y1,
        y2,
        selectionStyle,
        selectionComponent,
        name
      } = props;
      const width = Math.abs(x2 - x1) || 1;
      const height = Math.abs(y2 - y1) || 1;
      const x = Math.min(x1, x2);
      const y = Math.min(y1, y2);
      return y2 && x2 && x1 && y1 ? /*#__PURE__*/React.cloneElement(selectionComponent, {
        key: `${name}-selection`,
        x,
        y,
        width,
        height,
        style: selectionStyle
      }) : null;
    }

    // Overrides method in VictoryContainer
    getChildren(props) {
      return [...React.Children.toArray(props.children), this.getRect(props)];
    }
  };
}
export const VictorySelectionContainer = selectionContainerMixin(VictoryContainer);
