// @patternfly/react-core/dist/js/components/Text/Text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Text = exports.TextVariants = void 0;
const tslib_1 = require("tslib");
const React = tslib_1.__importStar(require("react"));
const react_styles_1 = require("@patternfly/react-styles");
const content_1 = tslib_1.__importDefault(require("@patternfly/react-styles/css/components/Content/content"));
const helpers_1 = require("../../helpers");
var TextVariants;
(function (TextVariants) {
    TextVariants["h1"] = "h1";
    TextVariants["h2"] = "h2";
    TextVariants["h3"] = "h3";
    TextVariants["h4"] = "h4";
    TextVariants["h5"] = "h5";
    TextVariants["h6"] = "h6";
    TextVariants["p"] = "p";
    TextVariants["a"] = "a";
    TextVariants["small"] = "small";
    TextVariants["blockquote"] = "blockquote";
    TextVariants["pre"] = "pre";
})(TextVariants || (exports.TextVariants = TextVariants = {}));
const componentStyles = {
    h1: content_1.default.contentH1,
    h2: content_1.default.contentH2,
    h3: content_1.default.contentH3,
    h4: content_1.default.contentH4,
    h5: content_1.default.contentH5,
    h6: content_1.default.contentH6,
    p: content_1.default.contentP,
    a: content_1.default.contentA,
    small: content_1.default.contentSmall,
    blockquote: content_1.default.contentBlockquote,
    pre: content_1.default.contentPre
};
const Text = (_a) => {
    var { children = null, className = '', component = TextVariants.p, isVisitedLink = false, ouiaId, ouiaSafe = true } = _a, props = tslib_1.__rest(_a, ["children", "className", "component", "isVisitedLink", "ouiaId", "ouiaSafe"]);
    const Component = component;
    const ouiaProps = (0, helpers_1.useOUIAProps)(exports.Text.displayName, ouiaId, ouiaSafe);
    return (React.createElement(Component, Object.assign({}, ouiaProps, props, { "data-pf-content": true, className: (0, react_styles_1.css)(isVisitedLink && component === TextVariants.a && content_1.default.modifiers.visited, componentStyles[component], className) }), children));
};
exports.Text = Text;
exports.Text.displayName = 'Text';
