// victory-line/lib/victory-line.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VictoryLine = void 0;
var _react = _interopRequireDefault(require("react"));
var _helperMethods = require("./helper-methods");
var _curve = require("./curve");
var _victoryCore = require("victory-core");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const fallbackProps = {
  width: 450,
  height: 300,
  padding: 50,
  interpolation: "linear"
};
const options = {
  components: [{
    name: "parent",
    index: "parent"
  }, {
    name: "data",
    index: "all"
  }, {
    name: "labels"
  }]
};

// eslint-disable-next-line @typescript-eslint/no-empty-interface

class VictoryLineBase extends _react.default.Component {
  constructor(props) {
    super(props);
  }
  static animationWhitelist = ["data", "domain", "height", "padding", "samples", "style", "width"];
  static displayName = "VictoryLine";
  static role = "line";
  static defaultTransitions = _victoryCore.DefaultTransitions.continuousTransitions();
  static defaultPolarTransitions = _victoryCore.DefaultTransitions.continuousPolarTransitions();
  static continuous = true;
  static defaultProps = {
    containerComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryContainer, null),
    dataComponent: /*#__PURE__*/_react.default.createElement(_curve.Curve, null),
    labelComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryLabel, {
      renderInPortal: true
    }),
    groupComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryClipContainer, null),
    samples: 50,
    sortKey: "x",
    sortOrder: "ascending",
    standalone: true,
    theme: _victoryCore.VictoryTheme.grayscale
  };
  static getDomain = _victoryCore.Domain.getDomain;
  static getData = _victoryCore.Data.getData;
  static getBaseProps(props) {
    return (0, _helperMethods.getBaseProps)(props, fallbackProps);
  }
  static expectedComponents = ["dataComponent", "labelComponent", "groupComponent", "containerComponent"];

  // Overridden in native versions
  shouldAnimate() {
    return !!this.props.animate;
  }
  render() {
    const {
      animationWhitelist,
      role
    } = VictoryLineBase;
    const props = _victoryCore.Helpers.modifyProps(this.props, fallbackProps, role);
    if (this.shouldAnimate()) {
      return this.animateComponent(props, animationWhitelist);
    }
    const children = this.renderContinuousData(props);
    const component = props.standalone ? this.renderContainer(props.containerComponent, children) : children;
    return _victoryCore.UserProps.withSafeUserProps(component, props);
  }
}
const VictoryLine = exports.VictoryLine = (0, _victoryCore.addEvents)(VictoryLineBase, options);
