// victory-core/lib/victory-portal/victory-portal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VictoryPortal = void 0;
var _defaults2 = _interopRequireDefault(require("lodash/defaults"));
var _react = _interopRequireDefault(require("react"));
var Log = _interopRequireWildcard(require("../victory-util/log"));
var Helpers = _interopRequireWildcard(require("../victory-util/helpers"));
var _portalContext = require("./portal-context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
class VictoryPortal extends _react.default.Component {
  static displayName = "VictoryPortal";
  static role = "portal";
  static defaultProps = {
    groupComponent: /*#__PURE__*/_react.default.createElement("g", null)
  };
  static contextType = _portalContext.PortalContext;
  componentDidMount() {
    if (!this.checkedContext) {
      if (typeof this.context.portalUpdate !== "function") {
        const msg = "`renderInPortal` is not supported outside of `VictoryContainer`. " + "Component will be rendered in place";
        Log.warn(msg);
        this.renderInPlace = true;
      }
      this.checkedContext = true;
    }
    this.forceUpdate();
  }
  componentDidUpdate() {
    if (!this.renderInPlace) {
      this.portalKey = this.portalKey || this.context.portalRegister();
      this.context.portalUpdate(this.portalKey, this.element);
    }
  }
  componentWillUnmount() {
    if (this.context && this.context.portalDeregister) {
      this.context.portalDeregister(this.portalKey);
    }
  }

  // Overridden in victory-core-native
  renderPortal(child) {
    if (this.renderInPlace) {
      return child;
    }
    this.element = child;
    return null;
  }
  render() {
    const children = Array.isArray(this.props.children) ? this.props.children[0] : this.props.children;
    const {
      groupComponent
    } = this.props;
    const childProps = children && children.props || {};
    const standardProps = childProps.groupComponent ? {
      groupComponent,
      standalone: false
    } : {};
    const newProps = (0, _defaults2.default)(standardProps, childProps, Helpers.omit(this.props, ["children", "groupComponent"]));
    const child = children && /*#__PURE__*/_react.default.cloneElement(children, newProps);
    return this.renderPortal(child);
  }
}
exports.VictoryPortal = VictoryPortal;
